// @antv/g2plot/lib/util/date.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.advanceBy = exports.getDay = exports.getWeek = exports.isLastDayOfMonth = exports.isLastWeekOfMonth = exports.getYearRange = exports.getDateRange = exports.DAY_MS = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var fecha_1 = tslib_1.__importDefault(require("fecha"));
var constant_1 = require("../plots/calendar/constant");
/**
 * 一天多少 ms
 */
exports.DAY_MS = 86400000;
/**
 * 获取最大最小日期范围
 * @param dates
 */
function getDateRange(dates) {
    var ds = tslib_1.__spreadArrays(dates).sort(function (a, b) { return a.getTime() - b.getTime(); });
    return [fecha_1.default.format(util_1.head(ds), constant_1.FORMATTER), fecha_1.default.format(util_1.last(ds), constant_1.FORMATTER)];
}
exports.getDateRange = getDateRange;
/**
 * 日期对应年的范围
 * @param date
 */
function getYearRange(date) {
    var curr = date ? date : new Date();
    return [
        fecha_1.default.format(new Date(curr.getFullYear(), 0, 1), constant_1.FORMATTER),
        fecha_1.default.format(new Date(curr.getFullYear(), 11, 30), constant_1.FORMATTER),
    ];
}
exports.getYearRange = getYearRange;
/**
 * 是否当前月的最后一周
 */
function isLastWeekOfMonth(date) {
    // 偏移 7 天之后，月份是否一样
    return date.getMonth() !== advanceBy(new Date(date), 7 * exports.DAY_MS).getMonth();
}
exports.isLastWeekOfMonth = isLastWeekOfMonth;
/**
 * 是否是当月的最后一天
 */
function isLastDayOfMonth(date) {
    // 偏移 1 天之后，月份是否一样
    return date.getMonth() !== advanceBy(new Date(date), exports.DAY_MS).getMonth();
}
exports.isLastDayOfMonth = isLastDayOfMonth;
/**
 * 获取 date 对应的周索引（国际标准：一年的第一个周四为第一周）
 * @param date
 */
function getWeek(date) {
    // 当年的第一天
    var oneJan = new Date(date.getFullYear(), 0, 1);
    return Math.ceil(((date.getTime() - oneJan.getTime()) / exports.DAY_MS + oneJan.getDay() + 1) / 7);
}
exports.getWeek = getWeek;
/**
 * 获得一周的第几天（周日第 0 天）
 * @param date
 */
function getDay(date) {
    return date.getDay();
}
exports.getDay = getDay;
/**
 * 将 Date 前进 ms 时间
 * @param d
 * @param ms
 */
function advanceBy(d, ms) {
    d.setMilliseconds(d.getMilliseconds() + ms);
    return d;
}
exports.advanceBy = advanceBy;
