// @antv/g2plot/lib/util/view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkShapeOverlap = exports.getOverlapArea = exports.moveInPanel = exports.checkOriginEqual = exports.forEachElement = exports.forEachGeometry = exports.getGeometryShapes = exports.getGeometryByType = exports.getAllGeometryByType = void 0;
var util_1 = require("@antv/util");
var dependents_1 = require("../dependents");
function getAllGeometryByType(view, type) {
    return util_1.filter(view.geometries, function (geometry) { return geometry.type === type; });
}
exports.getAllGeometryByType = getAllGeometryByType;
function getGeometryByType(view, type) {
    return util_1.head(getAllGeometryByType(view, type));
}
exports.getGeometryByType = getGeometryByType;
function getGeometryShapes(geometry) {
    return util_1.map(geometry.elements, function (element) { return element.shape; });
}
exports.getGeometryShapes = getGeometryShapes;
function forEachGeometry(view, callback) {
    view.geometries.forEach(callback);
}
exports.forEachGeometry = forEachGeometry;
function forEachElement(view, callback) {
    view.geometries.forEach(function (geometry) {
        geometry.elements.forEach(function (element) {
            callback(element);
        });
    });
}
exports.forEachElement = forEachElement;
/** 检测是否有和已存在的Shape数据`相等`的情况 */
function checkOriginEqual(cur, dones, compare) {
    return util_1.some(dones, function (done) {
        return compare(done.get(dependents_1.ORIGIN), cur.get(dependents_1.ORIGIN));
    });
}
exports.checkOriginEqual = checkOriginEqual;
/** 将label调整到panel内 */
function moveInPanel(shape, panel) {
    var box = shape.getBBox();
    var deltaX = 0;
    var deltaY = 0;
    if (box.minX < panel.minX) {
        deltaX = box.minX - panel.minX;
    }
    else if (box.maxX > panel.maxX) {
        deltaX = box.maxX - panel.maxX;
    }
    if (deltaX) {
        shape.attr('x', shape.attr('x') - deltaX);
    }
    if (box.minY < panel.minY) {
        deltaY = box.minY - panel.minY;
    }
    else if (box.maxY > panel.maxY) {
        deltaY = box.maxY - panel.maxY;
    }
    if (deltaY) {
        shape.attr('y', shape.attr('y') - deltaY);
    }
}
exports.moveInPanel = moveInPanel;
/**
 * 计算两个矩形之间的堆叠区域面积
 */
function getOverlapArea(a, b, margin) {
    if (margin === void 0) { margin = 0; }
    var xOverlap = Math.max(0, Math.min(a.x + a.width + margin, b.x + b.width + margin) - Math.max(a.x - margin, b.x - margin));
    var yOverlap = Math.max(0, Math.min(a.y + a.height + margin, b.y + b.height + margin) - Math.max(a.y - margin, b.y - margin));
    return xOverlap * yOverlap;
}
exports.getOverlapArea = getOverlapArea;
/** 检测是否和已布局的堆叠 */
function checkShapeOverlap(cur, dones) {
    var box = cur.getBBox();
    return util_1.some(dones, function (done) {
        var target = done.getBBox();
        return getOverlapArea(box, target, 2) > 0;
    });
}
exports.checkShapeOverlap = checkShapeOverlap;
