// @antv/g2plot/lib/plots/stacked-area/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../area/layer"));
var component_1 = require("./component");
require("./component/label/area-point");
require("./component/label/area-point-auto");
require("./theme");
var StackedAreaLayer = /** @class */ (function (_super) {
    tslib_1.__extends(StackedAreaLayer, _super);
    function StackedAreaLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.plotComponents = [];
        _this.type = 'stackedArea';
        _this.baseType = 'area';
        return _this;
    }
    StackedAreaLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            label: {
                visible: false,
                type: 'area',
            },
        });
    };
    StackedAreaLayer.prototype.beforeInit = function () {
        var visible = util_1.get(this.options, ['label', 'visible']);
        var type = util_1.get(this.options, ['label', 'type']);
        var options = this.options;
        if (visible) {
            if (type === 'line') {
                options.lineLabel = this.options.label;
            }
            if (type === 'area') {
                options.areaLabel = this.options.label;
            }
        }
        _super.prototype.beforeInit.call(this);
    };
    StackedAreaLayer.prototype.getColorScale = function () {
        var stackField = this.options.stackField;
        return this.view.getScaleByField(stackField);
    };
    StackedAreaLayer.prototype.label = function () {
        var props = this.options;
        var label = props.label;
        if (label && label.visible === false) {
            if (this.line) {
                this.line.label = false;
            }
            if (this.point) {
                this.point.label = false;
            }
            this.area.label = false;
            return;
        }
    };
    StackedAreaLayer.prototype.adjustArea = function (ele) {
        ele.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedAreaLayer.prototype.adjustLine = function (ele) {
        ele.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedAreaLayer.prototype.adjustPoint = function (ele) {
        ele.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedAreaLayer.prototype.afterRender = function () {
        this.renderPlotComponents();
        this.options.responsive = false;
        _super.prototype.afterRender.call(this);
    };
    StackedAreaLayer.prototype.geometryTooltip = function () {
        this.area.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.area.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.area.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.area.tooltip.fields = [this.options.xField, this.options.yField, this.options.stackField];
            }
        }
    };
    StackedAreaLayer.prototype.renderPlotComponents = function () {
        var _this = this;
        var componentsType = ['areaLabel', 'lineLabel'];
        util_1.each(componentsType, function (t) {
            var cfg = tslib_1.__assign({ view: _this.view, plot: _this }, _this.options[t]);
            var component = component_1.getPlotComponents(_this, t, cfg);
            if (component) {
                component.render();
                _this.plotComponents.push(component);
            }
        });
    };
    return StackedAreaLayer;
}(layer_1.default));
exports.default = StackedAreaLayer;
global_1.registerPlotType('stackedArea', StackedAreaLayer);
