// @antv/g2plot/esm/util/responsive/apply/axis.js
import { __extends } from "tslib";
import { each } from '@antv/util';
import ShapeNodes from '../node/shape-nodes';
import Responsive from '../responsive';
import ApplyResponsive from './base';
var SCALE_MAPPER = {
    cat: 'category',
    timeCat: 'category',
    time: 'dateTime',
    linear: 'linear',
};
var ApplyResponsiveAxis = /** @class */ (function (_super) {
    __extends(ApplyResponsiveAxis, _super);
    function ApplyResponsiveAxis() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ApplyResponsiveAxis.prototype.init = function () {
        this.axisInstance = this.getAxisInstance();
        _super.prototype.init.call(this);
    };
    ApplyResponsiveAxis.prototype.shouldApply = function () {
        var options = this.plot.options;
        if (!this.responsiveTheme.axis) {
            return false;
        }
        if (this.responsiveTheme.axis[this.dim] &&
            options[this.dim + "Axis"].visible &&
            options[this.dim + "Axis"].label &&
            options[this.dim + "Axis"].label.visible) {
            return true;
        }
        return false;
    };
    ApplyResponsiveAxis.prototype.apply = function () {
        var _this = this;
        var rawLabels = this.plot.view.backgroundGroup.findAll(function (el) {
            var name = el.get('name');
            if (name === 'axis-label') {
                var field = el.get('delegateObject').axis.get('field');
                if (field === _this.plot.options[_this.dim + "Field"]) {
                    return el;
                }
            }
        });
        var shapes = [];
        for (var i = 0; i < rawLabels.length; i++) {
            shapes.push(rawLabels[i]);
        }
        var shapeNodes = new ShapeNodes({
            shapes: shapes,
        });
        var _a = this.responsiveTheme.axis.x[this.type], constraints = _a.constraints, rules = _a.rules;
        new Responsive({
            nodes: shapeNodes,
            constraints: constraints,
            region: this.plot.getViewRange(),
            rules: rules,
            plot: this.plot,
            onEnd: function (nodes) {
                _this.updateTicks(nodes.origion_nodes);
            },
        });
    };
    ApplyResponsiveAxis.prototype.getType = function () {
        var props = this.plot.options;
        var axis = this.dim + "Axis";
        var field = this.dim + "Field";
        if (props[axis] && props[axis].type && props[axis].type === 'dateTime') {
            return 'dateTime';
        }
        var scaleType = this.plot.view.getScaleByField([props[field]]).type;
        return SCALE_MAPPER[scaleType];
    };
    ApplyResponsiveAxis.prototype.getAxisInstance = function () {
        var _a, _b;
        var axisIndex = this.dim === 'x' ? 0 : 1;
        var components = (_a = this.plot.view.getController('axis')) === null || _a === void 0 ? void 0 : _a.getComponents();
        if (components) {
            return (_b = components[axisIndex]) === null || _b === void 0 ? void 0 : _b.component;
        }
    };
    ApplyResponsiveAxis.prototype.updateTicks = function (nodes) {
        var _this = this;
        var tickLineContainer = this.plot.view.backgroundGroup.findAll(function (el) {
            var name = el.get('name');
            if (name === 'axis-tickline-group') {
                var field = el.get('delegateObject').axis.get('field');
                if (field === _this.plot.options[_this.dim + "Field"]) {
                    return el;
                }
            }
        })[0];
        if (tickLineContainer) {
            var tickShapes_1 = tickLineContainer.get('children');
            each(nodes, function (n, index) {
                if (n.shape.attr('text') === '') {
                    tickShapes_1[index].attr('opacity', 0);
                }
            });
        }
        this.plot.canvas.draw();
    };
    return ApplyResponsiveAxis;
}(ApplyResponsive));
export default ApplyResponsiveAxis;
