// @antv/g2plot/esm/util/responsive/apply/label.js
import { __extends } from "tslib";
import ShapeNodes from '../node/shape-nodes';
import Responsive from '../responsive';
import ApplyResponsive from './base';
import { each } from '@antv/util';
var ApplyResponsiveLabel = /** @class */ (function (_super) {
    __extends(ApplyResponsiveLabel, _super);
    function ApplyResponsiveLabel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ApplyResponsiveLabel.prototype.shouldApply = function () {
        if (!this.responsiveTheme.label || !this.responsiveTheme.label[this.type]) {
            return false;
        }
        return true;
    };
    ApplyResponsiveLabel.prototype.apply = function () {
        var labelShapesContainer = this.plot.view.geometries[0].labelsContainer.get('children');
        var labelShapes = [];
        each(labelShapesContainer, function (c) {
            labelShapes = labelShapes.concat(c.findAllByName('label'));
        });
        var nodes = new ShapeNodes({
            shapes: labelShapes,
        });
        var _a = this.responsiveTheme.label[this.type], constraints = _a.constraints, rules = _a.rules;
        new Responsive({
            nodes: nodes,
            constraints: constraints,
            rules: rules,
            plot: this.plot,
            region: this.plot.view.coordinateBBox,
        });
    };
    ApplyResponsiveLabel.prototype.getType = function () {
        return null;
    };
    return ApplyResponsiveLabel;
}(ApplyResponsive));
export default ApplyResponsiveLabel;
