// @antv/g2plot/lib/components/marker-point.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var dependents_1 = require("../dependents");
var util_1 = require("@antv/util");
var animate_1 = require("@antv/g2/lib/animate");
var DEFAULT_STYLE = {
    stroke: 'transparent',
    fill: '#FCC509',
    lineWidth: 0,
};
var ACTIVE_STYLE = {
    stroke: '#FFF',
    fill: '#FCC509',
    lineWidth: 1,
};
var SELECTED_STYLE = {
    stroke: 'rgba(0,0,0,0.85)',
    fill: '#FCC509',
    lineWidth: 1,
};
/**
 * 标注点 绘制在最顶层
 */
var MarkerPoint = /** @class */ (function () {
    function MarkerPoint(cfg) {
        this.points = [];
        this.labels = [];
        this.name = 'markerPoints';
        this.defaultCfg = {
            offsetX: 0,
            offsetY: 0,
            style: { normal: DEFAULT_STYLE, selected: SELECTED_STYLE, active: ACTIVE_STYLE },
            label: {
                visible: false,
                offsetY: -8,
                position: 'top',
                style: {
                    fill: 'rgba(0, 0, 0, 0.85)',
                },
            },
            animation: false,
        };
        this.view = cfg.view;
        this.size = cfg.size || 6;
        this.config = util_1.deepMix({}, this.defaultCfg, cfg);
        this._init();
    }
    MarkerPoint.prototype.render = function () {
        var dataArray = this.getDataArray();
        this._renderPoints(dataArray);
        this.view.canvas.draw();
        this._addInteraction();
    };
    MarkerPoint.prototype.clear = function () {
        if (this.container) {
            this.container.clear();
        }
    };
    MarkerPoint.prototype.destroy = function () {
        if (this.container) {
            this.container.remove();
        }
        this.points = [];
        this.labels = [];
    };
    MarkerPoint.prototype.getDataArray = function () {
        var geometry = this.view.geometries[0];
        return geometry.dataArray;
    };
    MarkerPoint.prototype._init = function () {
        var _this = this;
        var layer = this.view.foregroundGroup;
        this.container = layer.addGroup();
        this.render();
        this.view.on('beforerender', function () {
            _this.clear();
        });
    };
    MarkerPoint.prototype._renderPoints = function (dataArray) {
        var _this = this;
        util_1.each(this.config.data, function (dataItem, dataItemIdx) {
            util_1.each(dataArray, function (dataum) {
                var origin = util_1.find(dataum, function (d) { return util_1.isMatch(d._origin, dataItem); });
                if (origin) {
                    var pointAttrs = _this.config.style.normal;
                    var group = _this.container.addGroup({ name: _this.name });
                    var x = origin.x, y = origin.y;
                    if (util_1.isArray(x)) {
                        x = x[0];
                    }
                    if (util_1.isArray(y)) {
                        y = y[0];
                    }
                    var symbol = _this.config.symbol;
                    var _a = _this.config, offsetX = _a.offsetX, offsetY = _a.offsetY;
                    var point = void 0;
                    if (util_1.isString(symbol) && symbol.startsWith('image://')) {
                        var imageUrl = symbol.substr(8);
                        point = group.addShape('image', {
                            attrs: {
                                x: x - _this.size / 2 + offsetX,
                                y: y - _this.size / 2 + offsetY,
                                img: imageUrl,
                                width: _this.size,
                                height: _this.size,
                            },
                        });
                    }
                    else {
                        symbol = util_1.isString(symbol) ? dependents_1.MarkerSymbols[symbol] : symbol;
                        point = group.addShape({
                            type: 'marker',
                            name: 'marker-point',
                            id: "point-" + dataItemIdx,
                            attrs: tslib_1.__assign(tslib_1.__assign({ x: x + offsetX, y: y + offsetY, r: _this.size / 2 }, pointAttrs), { symbol: symbol }),
                        });
                    }
                    _this.points.push(point);
                    _this._animatePoint(point);
                    _this._renderLabel(group, origin, dataItemIdx);
                    group.set('data', dataItem);
                    group.set('origin', origin);
                }
            });
        });
    };
    MarkerPoint.prototype._renderLabel = function (container, origin, index) {
        var labelCfg = this.config.label;
        if (labelCfg && labelCfg.visible) {
            var _a = labelCfg.offsetX, offsetX = _a === void 0 ? 0 : _a, _b = labelCfg.offsetY, offsetY = _b === void 0 ? 0 : _b, formatter = labelCfg.formatter, position = labelCfg.position, field = labelCfg.field;
            var text = origin._origin[field];
            if (formatter) {
                text = formatter(text, { _origin: origin._origin }, index);
            }
            var x = util_1.isArray(origin.x) ? origin.x[0] : origin.x;
            var y = util_1.isArray(origin.y) ? origin.y[0] : origin.y;
            var label = container.addShape('text', {
                name: 'marker-label',
                id: "label-" + index,
                attrs: tslib_1.__assign(tslib_1.__assign({ x: x + offsetX, y: y + offsetY, text: text || '' }, labelCfg.style), { textAlign: 'center', textBaseline: position === 'top' ? 'bottom' : 'top' }),
            });
            this.labels.push(label);
        }
    };
    MarkerPoint.prototype._addInteraction = function () {
        var _this = this;
        var events = this.config.events;
        util_1.each(events, function (cb, eventName) {
            _this.container.on(_this.name + ":" + eventName, function (e) {
                cb(e);
                var target = e.target.get('parent');
                var pointShape = target.get('children')[0];
                if (pointShape) {
                    var data = pointShape.get('data');
                    var id = pointShape.get('id');
                    var condition = { id: id, data: data };
                    if (eventName === 'click') {
                        if (_this.selectedPoint && _this.selectedPoint.get('id') === id) {
                            _this.selectedPoint = null;
                            _this.setState('inactive', condition);
                        }
                        else {
                            _this.selectedPoint = pointShape;
                            _this.setState('selected', condition);
                        }
                    }
                    else if (eventName === 'mouseenter') {
                        _this.setState('active', condition);
                    }
                    else if (eventName === 'mouseleave') {
                        _this.setState('inactive', condition);
                    }
                }
                _this.view.canvas.draw();
            });
            _this.view.on('click', function (e) {
                var target = e.target.get('parent');
                if (!target || (target.get('name') !== _this.name && _this.selectedPoint)) {
                    _this.selectedPoint = null;
                    _this.setState('inactive');
                }
            });
        });
    };
    MarkerPoint.prototype.setState = function (state, condition) {
        var _this = this;
        if (state === 'active') {
            if (!this.selectedPoint || condition.id !== this.selectedPoint.get('id')) {
                this._onActive(condition);
            }
        }
        else if (state === 'inactive') {
            this.points.forEach(function (p) { return _this._onInactive(p); });
        }
        else if (state === 'selected') {
            this._onSelected(condition);
        }
    };
    MarkerPoint.prototype._onActive = function (condition) {
        var _this = this;
        var active = this.config.style.active;
        util_1.each(this.points, function (point) {
            if (point.get('id') === condition.id) {
                util_1.each(active, function (v, k) {
                    point.attr(k, v);
                });
            }
            else {
                _this._onInactive(point);
            }
        });
    };
    MarkerPoint.prototype._onInactive = function (point) {
        var normal = this.config.style.normal;
        if (!this.selectedPoint || point.get('id') !== this.selectedPoint.get('id')) {
            util_1.each(normal, function (v, k) {
                point.attr(k, v);
            });
        }
    };
    MarkerPoint.prototype._onSelected = function (condition) {
        var _this = this;
        var selected = this.config.style.selected;
        util_1.each(this.points, function (point) {
            if (point.get('id') === condition.id) {
                util_1.each(selected, function (v, k) {
                    point.attr(k, v);
                });
            }
            else {
                _this._onInactive(point);
            }
        });
    };
    /** point animation, not for label */
    MarkerPoint.prototype._animatePoint = function (shape) {
        var _a = this.config, animation = _a.animation, size = _a.size;
        if (animation !== false) {
            var _b = animation, _c = _b.endState, endState = _c === void 0 ? {} : _c, _d = _b.animateCfg, animateCfg = _d === void 0 ? animate_1.DEFAULT_ANIMATE_CFG.appear : _d;
            shape.animate(tslib_1.__assign({ r: Number.isNaN(endState.size / 2) ? size / 2 : endState.size / 2 }, endState), animateCfg);
        }
    };
    return MarkerPoint;
}());
exports.default = MarkerPoint;
