// @antv/g2plot/lib/components/statistic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("@antv/util");
var dom_util_1 = require("@antv/dom-util");
var StatisticHtml = /** @class */ (function () {
    function StatisticHtml(cfg) {
        var defaultOptions = this.getDefaultOptions();
        this.options = util_1.deepMix(defaultOptions, cfg, {});
        this.x = this.options.x;
        this.y = this.options.y;
        this.html = this.options.html;
        this.container = this.options.container;
    }
    StatisticHtml.prototype.render = function () {
        if (util_1.isElement(this.container)) {
            this.wrapperNode = dom_util_1.createDom('<div class="g2plot-htmlStatistic"></div>');
            this.container.appendChild(this.wrapperNode);
            dom_util_1.modifyCSS(this.wrapperNode, {
                position: 'absolute',
                pointerEvents: 'none',
            });
            var htmlNode = this.html;
            this.wrapperNode.appendChild(htmlNode);
            this.setDomPosition(this.x, this.y);
        }
    };
    StatisticHtml.prototype.updateHtml = function (content) {
        if (util_1.isType(content, 'HTMLDivElement')) {
            this.wrapperNode.innerHTML = '';
            this.wrapperNode.appendChild(content);
        }
        else {
            this.wrapperNode.innerHTML = content;
        }
        this.setDomPosition(this.x, this.y);
    };
    StatisticHtml.prototype.updatePosition = function (x, y) {
        this.x = x;
        this.y = y;
        this.setDomPosition(x, y);
    };
    StatisticHtml.prototype.destroy = function () {
        this.container.removeChild(this.wrapperNode);
    };
    StatisticHtml.prototype.getDefaultOptions = function () {
        return {
            x: 0,
            y: 0,
            width: 0,
            height: 0,
            html: '',
            container: null,
            alignX: 'middle',
            alignY: 'middle',
        };
    };
    StatisticHtml.prototype.setDomPosition = function (x, y) {
        var xPosition = x;
        var yPosition = y;
        var width = dom_util_1.getOuterWidth(this.wrapperNode);
        var height = dom_util_1.getOuterHeight(this.wrapperNode);
        if (this.options.alignX === 'middle') {
            xPosition = x - width / 2;
        }
        if (this.options.alignY === 'middle') {
            yPosition = y - height / 2;
        }
        dom_util_1.modifyCSS(this.wrapperNode, {
            top: Math.round(yPosition) + "px",
            left: Math.round(xPosition) + "px",
            /**
             * 支持Ant Design Charts
             * render是异步的操作，获取到的 width 为 0
             */
            transform: width > 0 ? 'none' : 'translate(-50%, -50%)',
        });
    };
    return StatisticHtml;
}());
exports.default = StatisticHtml;
