// @antv/g2plot/lib/plots/gauge/geometry/shape/gauge-shape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GaugeShape = void 0;
var tslib_1 = require("tslib");
/**
 * @author linhuiw
 * @description 仪表盘形状
 */
var util_1 = require("@antv/util");
var g2_1 = require("@antv/g2");
var theme_1 = require("../../../../theme");
var common_1 = require("../../../../util/common");
/**
 * 仪表盘指针图形
 * 指针主体由梯形和一大一小圆形组成，
 * 中心处由灰色圆底和小白圆加以装饰
 */
var GaugeShape = /** @class */ (function () {
    function GaugeShape(uid) {
        this.uid = uid;
    }
    GaugeShape.prototype.setOption = function (type, options) {
        this.type = type;
        this.options = options;
        this.axis = options.axis;
        this.pivot = options.pivot;
    };
    GaugeShape.prototype.render = function () {
        var Gauge = this; // eslint-disable-line @typescript-eslint/no-this-alias
        g2_1.registerShape('point', 'gauge', {
            draw: function (cfg, group) {
                var _a, _b, _c;
                this.gauge = {};
                this.gauge.options = Gauge.options;
                this.gauge.axis = Gauge.axis;
                this.gauge.pivot = Gauge.pivot;
                this.gauge.type = Gauge.type;
                var gauge = this.gauge;
                var type = this.gauge.type;
                var point = cfg.points[0];
                var center = this.parsePoint({
                    x: 0,
                    y: 0,
                });
                var target = this.parsePoint({
                    x: point.x || 0,
                    y: 1,
                });
                gauge.center = center;
                gauge.group = group;
                var r = { x: center.x - target.x, y: center.y - target.y };
                this.gauge.ringRadius = Math.sqrt(r.x * r.x + r.y * r.y);
                var _d = this.getAngleRange(), starAngle = _d.starAngle, endAngle = _d.endAngle;
                var currentAngle = point.x * (endAngle - starAngle) + starAngle;
                switch (type) {
                    case 'meterGauge':
                        this.drawBarGauge(currentAngle);
                        if (this.gauge.axis.visible && ((_a = this.gauge.axis.tickLine) === null || _a === void 0 ? void 0 : _a.visible)) {
                            this.drawInSideAxis();
                        }
                        break;
                    case 'fanGauge':
                        this.drawGauge(currentAngle);
                        if (this.gauge.axis.visible && ((_b = this.gauge.axis.tickLine) === null || _b === void 0 ? void 0 : _b.visible)) {
                            this.drawOutSideAxis();
                        }
                        break;
                    case 'standardGauge':
                    default:
                        this.drawGauge(currentAngle);
                        if (this.gauge.axis.visible && ((_c = this.gauge.axis.tickLine) === null || _c === void 0 ? void 0 : _c.visible)) {
                            this.drawAxis();
                        }
                        break;
                }
                // 绘制指针
                if (this.gauge.pivot.visible) {
                    this.drawPivot(cfg, group);
                }
            },
            drawGauge: function (currentAngle) {
                var range = this.gauge.options.range;
                this.drawBottomRing(); // 绘制灰底色
                if (range && range.length) {
                    this.drawRangeColor();
                }
                else {
                    this.drawCurrentRing(currentAngle);
                }
            },
            drawRangeColor: function () {
                var _a = this.gauge.options, min = _a.min, max = _a.max, range = _a.range, color = _a.color, rangeStyle = _a.rangeStyle;
                var colors = color || theme_1.getGlobalTheme().colors;
                var _b = this.getAngleRange(), starAngle = _b.starAngle, endAngle = _b.endAngle;
                var config = {
                    min: min,
                    max: max,
                    starAngle: starAngle,
                    endAngle: endAngle,
                };
                for (var i = 0; i < range.length; i++) {
                    var start = this.valueToAngle(range[i], config);
                    var end = this.valueToAngle(range[i + 1], config);
                    if (end >= start) {
                        var path2 = this.getPath(start, end);
                        var style = util_1.deepMix({ fill: colors[i] }, rangeStyle);
                        this.drawRing(path2, style);
                    }
                }
            },
            drawBottomRing: function () {
                var _a = this.getAngleRange(), starAngle = _a.starAngle, endAngle = _a.endAngle;
                var backgroundStyle = this.gauge.options.rangeBackgroundStyle;
                var path = this.getPath(starAngle, endAngle);
                this.drawRing(path, backgroundStyle);
            },
            drawCurrentRing: function (current) {
                var starAngle = this.getAngleRange().starAngle;
                var rangeStyle = this.gauge.rangeStyle;
                var path3 = this.getPath(starAngle, current);
                this.drawRing(path3, rangeStyle);
            },
            drawInSideAxis: function () {
                var _a = this.gauge.options, min = _a.min, max = _a.max, axis = _a.axis;
                var _b = this.getAngleRange(), starAngle = _b.starAngle, endAngle = _b.endAngle;
                var config = {
                    min: min,
                    max: max,
                    starAngle: starAngle,
                    endAngle: endAngle,
                };
                var interval = (max - min) / axis.tickCount;
                for (var i = 0; i < axis.tickCount; i++) {
                    var startValue = min + i * interval;
                    var angle = this.valueToAngle(startValue + interval / 2, config);
                    this.drawRect(angle, {
                        length: axis.tickLine.length,
                        style: axis.tickLine.style,
                    });
                }
            },
            drawAxis: function () {
                var axis = this.gauge.axis;
                var _a = this.gauge.options, min = _a.min, max = _a.max;
                var _b = this.getAngleRange(), starAngle = _b.starAngle, endAngle = _b.endAngle;
                var config = {
                    min: min,
                    max: max,
                    starAngle: starAngle,
                    endAngle: endAngle,
                };
                var interval = (max - min) / (axis.tickCount - 1);
                for (var i = 0; i < axis.tickCount; i++) {
                    var startValue = min + i * interval;
                    var angle = this.valueToAngle(startValue, config);
                    var tickLineStyle = util_1.clone(axis.tickLine.style);
                    if (i % 5 !== 0) {
                        tickLineStyle.lineWidth = tickLineStyle.lineWidth / 2;
                    }
                    this.drawRect(angle, {
                        length: i % 5 === 0 ? axis.tickLine.length : axis.tickLine.length / 2,
                        style: tickLineStyle,
                    });
                }
            },
            drawOutSideAxis: function () {
                var axis = this.gauge.axis;
                var _a = this.gauge.options, min = _a.min, max = _a.max;
                var _b = this.getAngleRange(), starAngle = _b.starAngle, endAngle = _b.endAngle;
                var config = {
                    min: min,
                    max: max,
                    starAngle: starAngle,
                    endAngle: endAngle,
                };
                var interval = (max - min) / (axis.tickCount - 1);
                for (var i = 0; i < axis.tickCount; i++) {
                    var startValue = min + i * interval;
                    var angle = this.valueToAngle(startValue, config);
                    this.drawRect(angle, {
                        length: axis.tickLine.length,
                        style: axis.tickLine.style,
                    });
                }
            },
            drawBarGauge: function (current) {
                var _this = this;
                var _a = this.gauge.options, min = _a.min, max = _a.max, range = _a.range, color = _a.color, rangeStyle = _a.rangeStyle, rangeBackgroundStyle = _a.rangeBackgroundStyle;
                var colors = color || theme_1.getGlobalTheme().colors;
                var _b = this.getAngleRange(), starAngle = _b.starAngle, endAngle = _b.endAngle;
                var config = {
                    min: min,
                    max: max,
                    starAngle: starAngle,
                    endAngle: endAngle,
                };
                var interval = (endAngle - starAngle) / (50 - 1);
                var offset = interval / 3;
                // 由50个柱子组成
                for (var i = 0; i < 50; i++) {
                    var start = starAngle + i * interval;
                    var path2 = this.getPath(start - offset / 2, start + offset - offset / 2);
                    var style = rangeBackgroundStyle;
                    if (range && range.length) {
                        var result1 = range.map(function (item) {
                            return _this.valueToAngle(item, config);
                        });
                        var index = common_1.sortedLastIndex(result1, start);
                        /** 最后一个值也在最后一个区间内 */
                        var colorIndex = Math.min(index, range.length - 1);
                        style = util_1.deepMix({}, { fill: colors[colorIndex - 1] }, rangeStyle) || rangeBackgroundStyle;
                    }
                    else {
                        style = current >= start ? util_1.deepMix({}, { fill: color }, rangeStyle) : rangeBackgroundStyle;
                    }
                    this.drawRing(path2, style);
                }
            },
            getAngleRange: function () {
                var angle = this.gauge.options.angle;
                var angleValue = 90 - (360 - angle) * 0.5;
                var starAngle = ((270 - 90 - angleValue) * Math.PI) / 180;
                var endAngle = ((270 + 90 + angleValue) * Math.PI) / 180;
                return { starAngle: starAngle, endAngle: endAngle };
            },
            valueToAngle: function (value, config) {
                var min = config.min, max = config.max, starAngle = config.starAngle, endAngle = config.endAngle;
                if (value === max) {
                    return endAngle;
                }
                if (value === min) {
                    return starAngle;
                }
                var ratio = (value - min) / (max - min);
                if (max === min) {
                    ratio = 1;
                }
                var angle = ratio * (endAngle - starAngle) + starAngle;
                angle = Math.max(angle, starAngle);
                angle = Math.min(angle, endAngle);
                return angle;
            },
            drawRing: function (path, style) {
                this.gauge.group.addShape('path', {
                    attrs: util_1.deepMix({}, {
                        path: path,
                    }, style),
                });
            },
            drawRect: function (angle, param) {
                var axis = this.gauge.axis;
                var config = tslib_1.__assign(tslib_1.__assign({}, axis), param);
                var offset = config.offset, length = config.length;
                var center = this.gauge.center;
                var radius;
                if (offset < 0) {
                    radius = this.gauge.ringRadius - this.gauge.options.rangeSize + offset;
                }
                else {
                    radius = this.gauge.ringRadius + offset;
                }
                var xA1 = radius * Math.cos(angle) + center.x;
                var yA1 = radius * Math.sin(angle) + center.y;
                var xB1 = (radius + length) * Math.cos(angle) + center.x;
                var yB1 = (radius + length) * Math.sin(angle) + center.y;
                var line = this.gauge.group.addShape('line', {
                    attrs: util_1.deepMix({}, {
                        x1: xA1,
                        y1: yA1,
                        x2: xB1,
                        y2: yB1,
                    }, param.style),
                });
                line.set('name', 'axis-tickLine');
            },
            getPath: function (starAngle, endAngle) {
                /* const type = this.gauge.type;
                const height = get(gauge, 'options.height');
                const width = get(gauge, 'options.width');*/
                var center = this.gauge.center;
                var length = this.gauge.ringRadius;
                /*let thick;
                 const { minThickness, minThickCanvsSize } = this.gauge.ringStyle;
                const size = Math.min(width, height);
                if (type === 'fan' && size < minThickCanvsSize) {
                  thick = length - minThickness;
                } else {*/
                var thick = this.gauge.options.rangeSize;
                //}
                var xA1 = length * Math.cos(starAngle) + center.x;
                var yA1 = length * Math.sin(starAngle) + center.y;
                var xA2 = (length - thick) * Math.cos(starAngle) + center.x;
                var yA2 = (length - thick) * Math.sin(starAngle) + center.y;
                var xB1 = length * Math.cos(endAngle) + center.x;
                var yB1 = length * Math.sin(endAngle) + center.y;
                var xB2 = (length - thick) * Math.cos(endAngle) + center.x;
                var yB2 = (length - thick) * Math.sin(endAngle) + center.y;
                var largeArcFlag = Math.abs(starAngle - endAngle) > Math.PI ? 1 : 0;
                return [
                    ['M', xA1, yA1],
                    ['A', length, length, 0, largeArcFlag, 1, xB1, yB1],
                    ['L', xB2, yB2],
                    ['A', length - thick, length - thick, 0, largeArcFlag, 0, xA2, yA2],
                    ['Z'],
                ];
            },
            drawPivot: function (cfg) {
                var _a = this.getAngleRange(), starAngle = _a.starAngle, endAngle = _a.endAngle;
                var _b = this.gauge.options, radius = _b.radius, rangeSize = _b.rangeSize;
                var pivotConfig = this.gauge.pivot;
                var bigCircle = pivotConfig.thickness;
                var smCircle = pivotConfig.thickness / 2.5;
                var group = this.gauge.group;
                var point = cfg.points[0];
                var center = this.parsePoint({
                    x: 0,
                    y: 0,
                });
                // radius
                var current = point.x * (endAngle - starAngle) + starAngle;
                var x = (this.gauge.ringRadius - rangeSize) * radius * Math.cos(current) + this.gauge.center.x;
                var y = (this.gauge.ringRadius - rangeSize) * radius * Math.sin(current) + this.gauge.center.y;
                var target = {
                    x: x,
                    y: y,
                };
                // 外底色灰圆
                if (pivotConfig.base.visible) {
                    group.addShape('circle', {
                        attrs: util_1.deepMix({}, {
                            x: center.x,
                            y: center.y,
                            r: pivotConfig.base.size ? pivotConfig.base.size / 2 : bigCircle * 2.2,
                        }, pivotConfig.base.style),
                    });
                }
                // 指针
                if (pivotConfig.pointer.visible) {
                    var dirVec = { x: center.x - target.x, y: center.y - target.y };
                    var length_1 = Math.sqrt(dirVec.x * dirVec.x + dirVec.y * dirVec.y);
                    dirVec.x *= 1 / length_1;
                    dirVec.y *= 1 / length_1;
                    var angle1 = -Math.PI / 2;
                    var x1 = Math.cos(angle1) * dirVec.x - Math.sin(angle1) * dirVec.y;
                    var y1 = Math.sin(angle1) * dirVec.x + Math.cos(angle1) * dirVec.y;
                    var angle2 = Math.PI / 2;
                    var x2 = Math.cos(angle2) * dirVec.x - Math.sin(angle2) * dirVec.y;
                    var y2 = Math.sin(angle2) * dirVec.x + Math.cos(angle2) * dirVec.y;
                    var path = [
                        ['M', target.x + x1 * smCircle, target.y + y1 * smCircle],
                        ['L', center.x + x1 * bigCircle, center.y + y1 * bigCircle],
                        ['L', center.x + x2 * bigCircle, center.y + y2 * bigCircle],
                        ['L', target.x + x2 * smCircle, target.y + y2 * smCircle],
                        ['Z'],
                    ];
                    group.addShape('path', {
                        attrs: util_1.deepMix({}, {
                            path: path,
                        }, pivotConfig.pointer.style),
                    });
                    group.addShape('circle', {
                        attrs: {
                            x: target.x,
                            y: target.y,
                            r: smCircle,
                            fill: pivotConfig.pointer.style.fill,
                        },
                    });
                    group.addShape('circle', {
                        attrs: {
                            x: center.x,
                            y: center.y,
                            r: bigCircle,
                            fill: pivotConfig.pointer.style.fill,
                        },
                    });
                }
                if (pivotConfig.pin.visible) {
                    // 内部白色小圆
                    group.addShape('circle', {
                        attrs: util_1.deepMix({}, {
                            x: center.x,
                            y: center.y,
                            r: smCircle / 1.2,
                        }, pivotConfig.pin.style),
                    });
                }
            },
        });
    };
    return GaugeShape;
}());
exports.GaugeShape = GaugeShape;
