// @antv/g2plot/lib/plots/funnel/animation/funnel-scale-in-y.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var matrix_util_1 = require("@antv/matrix-util");
var g2_1 = require("@antv/g2");
function funnelScaleInY(shape, animateCfg) {
    var _a = animateCfg || {}, _b = _a.duration, duration = _b === void 0 ? 200 : _b, delay = _a.delay, easing = _a.easing, callback = _a.callback, reverse = _a.reverse;
    var bbox = shape.getBBox();
    var originX = (bbox.minX + bbox.maxX) / 2;
    var originY = reverse ? bbox.maxY : bbox.minY;
    var clip = shape.setClip({
        type: 'rect',
        attrs: {
            x: bbox.x,
            y: bbox.y,
            width: bbox.width,
            height: bbox.height,
        },
    });
    var clipTargetAttrs = {
        matrix: [1, 0, 0, 0, 1, 0, 0, 0, 1],
    };
    clip.setMatrix(matrix_util_1.ext.transform(clip.getMatrix(), [
        ['t', -originX, -originY],
        ['s', 1, 0],
        ['t', originX, originY],
    ]));
    var shapeTargetAttrs = {
        fillOpacity: shape.attr('fillOpacity'),
        strokeOpacity: shape.attr('strokeOpacity'),
        opacity: shape.attr('opacity'),
    };
    shape.attr({
        fillOpacity: 0,
        strokeOpacity: 0,
        opacity: 0,
    });
    clip.animate(clipTargetAttrs, {
        duration: 200,
        easing: easing,
        callback: function () {
            shape.setClip(null);
            clip.remove();
        },
        delay: delay,
    });
    shape.animate(shapeTargetAttrs, { duration: duration, easing: easing, delay: delay });
    callback && setTimeout(function () { return callback(shape); }, duration + delay);
}
funnelScaleInY.animationName = 'funnelScaleInY';
g2_1.registerAnimation('funnelScaleInY', funnelScaleInY);
