// @antv/g2plot/lib/plots/radar/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * Create By Bruce Too
 * On 2020-02-14
 */
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var view_layer_1 = tslib_1.__importDefault(require("../../base/view-layer"));
var factory_1 = require("../../components/factory");
var factory_2 = require("../../geoms/factory");
var scale_1 = require("../../util/scale");
var EventParser = tslib_1.__importStar(require("./event"));
require("./theme");
var GEOM_MAP = {
    area: 'area',
    line: 'line',
    point: 'point',
};
var RadarLayer = /** @class */ (function (_super) {
    tslib_1.__extends(RadarLayer, _super);
    function RadarLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'radar';
        return _this;
    }
    RadarLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            width: 400,
            height: 400,
            title: {
                visible: false,
            },
            description: {
                visible: false,
            },
            forceFit: true,
            padding: 'auto',
            radius: 0.8,
            smooth: false,
            line: {
                visible: true,
                size: 2,
                style: {
                    opacity: 1,
                },
            },
            area: {
                visible: true,
                style: {
                    opacity: 0.25,
                },
            },
            point: {
                visible: false,
                size: 4,
                shape: 'point',
                style: {
                    opacity: 1,
                },
            },
            angleAxis: {
                visible: true,
                autoRotateTitle: true,
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                grid: {
                    visible: true,
                    line: {
                        style: {
                            lineDash: [0, 0],
                        },
                    },
                },
                label: {
                    visible: true,
                    offset: 16,
                    autoRotate: true,
                    autoHide: true,
                },
                title: {
                    visible: false,
                },
            },
            radiusAxis: {
                min: 0,
                visible: true,
                /** G2 4.0 默认 nice 不生效，需要手动添加 */
                nice: true,
                autoRotateTitle: true,
                line: {
                    visible: true,
                },
                tickLine: {
                    visible: true,
                },
                gridType: 'line',
                grid: {
                    visible: true,
                    line: {
                        style: {
                            lineDash: [0, 0],
                        },
                    },
                },
                label: {
                    visible: true,
                    autoHide: true,
                    autoRotate: true,
                },
                title: {
                    visible: false,
                },
            },
            label: {
                visible: false,
                type: 'point',
            },
            legend: {
                visible: true,
                position: 'left-top',
            },
            tooltip: {
                visible: true,
                shared: true,
                showCrosshairs: false,
            },
        });
    };
    RadarLayer.prototype.init = function () {
        var props = this.options;
        props.xField = props.angleField;
        props.yField = props.radiusField;
        _super.prototype.init.call(this);
    };
    RadarLayer.prototype.getAngleScale = function () {
        var angleField = this.options.angleField;
        return this.view.getScaleByField(angleField);
    };
    RadarLayer.prototype.getRadiusScale = function () {
        var radiusField = this.options.radiusField;
        return this.view.getScaleByField(radiusField);
    };
    RadarLayer.prototype.getColorScale = function () {
        var seriesField = this.options.seriesField;
        return this.view.getScaleByField(seriesField);
    };
    RadarLayer.prototype.geometryParser = function (dim, type) {
        return GEOM_MAP[type];
    };
    RadarLayer.prototype.scale = function () {
        var props = this.options;
        var scales = {};
        /** 配置x-scale */
        scales[props.angleField] = {};
        if (util_1.has(props, 'angleAxis')) {
            scale_1.extractScale(scales[props.angleField], props.angleAxis);
        }
        /** 配置y-scale */
        scales[props.radiusField] = {};
        if (util_1.has(props, 'radiusAxis')) {
            scale_1.extractScale(scales[props.radiusField], props.radiusAxis);
        }
        this.setConfig('scales', scales);
        _super.prototype.scale.call(this);
    };
    RadarLayer.prototype.coord = function () {
        var props = this.options;
        var coordConfig = {
            type: 'polar',
            cfg: {
                radius: props.radius,
            },
        };
        this.setConfig('coordinate', coordConfig);
    };
    RadarLayer.prototype.axis = function () {
        var props = this.options;
        var xAxis_parser = factory_1.getComponent('axis', {
            plot: this,
            dim: 'angle',
        });
        var yAxis_parser = factory_1.getComponent('axis', {
            plot: this,
            dim: 'radius',
        });
        var axesConfig = {};
        axesConfig[props.angleField] = xAxis_parser;
        axesConfig[props.radiusField] = yAxis_parser;
        /** 存储坐标轴配置项到config */
        this.setConfig('axes', axesConfig);
    };
    RadarLayer.prototype.addGeometry = function () {
        var props = this.options;
        /** 配置面积 */
        if (props.area.visible) {
            var area = factory_2.getGeom('area', 'main', {
                plot: this,
            });
            this.setConfig('geometry', area);
            this.area = area;
        }
        /** 配置线 */
        if (props.line && props.line.visible) {
            var line = factory_2.getGeom('line', 'main', {
                plot: this,
            });
            this.setConfig('geometry', line);
            this.line = line;
        }
        /** 配置点 */
        if (props.point && props.point.visible) {
            var point = factory_2.getGeom('point', 'guide', {
                plot: this,
            });
            this.setConfig('geometry', point);
            this.point = point;
        }
        if (props.label) {
            this.label();
        }
        if (props.tooltip && (props.tooltip.fields || props.tooltip.formatter)) {
            this.geometryTooltip();
        }
    };
    RadarLayer.prototype.geometryTooltip = function () {
        var geomConfig = this.line ? this.line : this.area;
        geomConfig.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            geomConfig.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            geomConfig.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                geomConfig.tooltip.fields = [this.options.angleField, this.options.radiusField];
            }
            if (this.options.seriesField) {
                geomConfig.tooltip.fields.push(this.options.seriesField);
            }
        }
    };
    RadarLayer.prototype.label = function () {
        var props = this.options;
        if (props.label.visible === false) {
            if (this.point) {
                this.point.label = false;
            }
            if (this.line) {
                this.line.label = false;
            }
            if (this.area) {
                this.area.label = false;
            }
            return;
        }
        // @Todo 雷达图标签布局算法后续补充
        var label = factory_1.getComponent('label', tslib_1.__assign({ fields: [props.radiusField], cfg: {
                type: 'polar',
                autoRotate: false,
            }, plot: this }, props.label));
        if (this.point) {
            this.point.label = label;
        }
        else if (this.line) {
            this.line.label = label;
        }
        else if (this.area) {
            this.area.label = label;
        }
    };
    RadarLayer.prototype.annotation = function () {
        return;
    };
    RadarLayer.prototype.animation = function () {
        _super.prototype.animation.call(this);
        var props = this.options;
        if (props.animation === false) {
            // 关闭动画
            if (this.area)
                this.area.animate = false;
            if (this.line)
                this.line.animate = false;
            if (this.point)
                this.point.animate = false;
        }
    };
    RadarLayer.prototype.parseEvents = function () {
        _super.prototype.parseEvents.call(this, EventParser);
    };
    return RadarLayer;
}(view_layer_1.default));
exports.default = RadarLayer;
global_1.registerPlotType('radar', RadarLayer);
