// @antv/g2plot/esm/plots/percent-stacked-area/layer.js
import { __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import StackedArea from '../stacked-area/layer';
import { transformDataPercentage } from '../../util/data';
var PercentStackedAreaLayer = /** @class */ (function (_super) {
    __extends(PercentStackedAreaLayer, _super);
    function PercentStackedAreaLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'percentStackedArea';
        _this.baseType = 'stackedArea';
        return _this;
    }
    PercentStackedAreaLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            yAxis: {
                visible: true,
                label: {
                    visible: true,
                    formatter: function (v) {
                        var reg = /%/gi;
                        return v.replace(reg, '');
                    },
                },
            },
        });
    };
    PercentStackedAreaLayer.prototype.processData = function (originData) {
        var _a = this.options, xField = _a.xField, yField = _a.yField;
        return transformDataPercentage(originData, xField, [yField]);
    };
    PercentStackedAreaLayer.prototype.scale = function () {
        var metaConfig = {};
        var yField = this.options.yField;
        metaConfig[this.options.yField] = {
            tickCount: 6,
            alias: yField + " (%)",
            min: 0,
            max: 1,
            formatter: function (v) {
                var formattedValue = (v * 100).toFixed(1);
                return formattedValue + "%";
            },
        };
        this.options.meta = deepMix({}, metaConfig, this.options.meta);
        _super.prototype.scale.call(this);
    };
    return PercentStackedAreaLayer;
}(StackedArea));
export default PercentStackedAreaLayer;
registerPlotType('percentStackedArea', PercentStackedAreaLayer);
