// @antv/g2plot/lib/plots/percent-stacked-bar/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../stacked-bar/layer"));
var data_1 = require("../../util/data");
var PercentStackedBarLayer = /** @class */ (function (_super) {
    tslib_1.__extends(PercentStackedBarLayer, _super);
    function PercentStackedBarLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'percentStackedBar';
        return _this;
    }
    PercentStackedBarLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            xAxis: {
                visible: true,
                tickLine: {
                    visible: false,
                },
                grid: {
                    visible: false,
                },
                title: {
                    visible: true,
                    formatter: function (v) { return v + " (%)"; },
                },
                label: {
                    visible: false,
                    formatter: function (v) {
                        var reg = /%/gi;
                        return v.replace(reg, '');
                    },
                },
            },
        });
    };
    PercentStackedBarLayer.prototype.processData = function (originData) {
        var _a = this.options, xField = _a.xField, yField = _a.yField;
        var processData = _super.prototype.processData.call(this, originData);
        return data_1.transformDataPercentage(processData, yField, [xField]);
    };
    PercentStackedBarLayer.prototype.scale = function () {
        var metaConfig = {};
        var xField = this.options.xField;
        metaConfig[xField] = {
            tickCount: 6,
            alias: xField + " (%)",
            min: 0,
            max: 1,
            formatter: function (v) {
                var formattedValue = (v * 100).toFixed(1);
                return formattedValue + "%";
            },
        };
        this.options.meta = util_1.deepMix({}, this.options.meta, metaConfig);
        _super.prototype.scale.call(this);
    };
    return PercentStackedBarLayer;
}(layer_1.default));
exports.default = PercentStackedBarLayer;
global_1.registerPlotType('percentStackedBar', PercentStackedBarLayer);
