// @antv/g2/lib/geometry/shape/interval/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var base_1 = require("../base");
var get_style_1 = require("../util/get-style");
var util_1 = require("./util");
/** Interval 的 shape 工厂 */
var IntervalShapeFactory = base_1.registerShapeFactory('interval', {
    defaultShapeType: 'rect',
    getDefaultPoints: function (pointInfo) {
        return util_1.getRectPoints(pointInfo);
    },
});
/** Inerval 默认 shape，填充的矩形 */
base_1.registerShape('interval', 'rect', {
    draw: function (cfg, container) {
        var style = get_style_1.getStyle(cfg, false, true);
        var path = this.parsePath(util_1.getRectPath(cfg.points));
        var shape = container.addShape('path', {
            attrs: tslib_1.__assign(tslib_1.__assign({}, style), { path: path }),
            name: 'interval',
        });
        return shape;
    },
    getMarker: function (markerCfg) {
        var color = markerCfg.color, isInPolar = markerCfg.isInPolar;
        if (isInPolar) {
            return {
                symbol: 'circle',
                style: {
                    r: 4.5,
                    fill: color,
                },
            };
        }
        return {
            symbol: 'square',
            style: {
                r: 4,
                fill: color,
            },
        };
    },
});
exports.default = IntervalShapeFactory;
