// @antv/g2plot/lib/plots/line/interaction/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineSelect = exports.LineActive = void 0;
var tslib_1 = require("tslib");
var line_active_1 = tslib_1.__importDefault(require("./line-active"));
exports.LineActive = line_active_1.default;
var line_select_1 = tslib_1.__importDefault(require("./line-select"));
exports.LineSelect = line_select_1.default;
var dependents_1 = require("../../../dependents");
var selected_tooltip_1 = require("./selected-tooltip");
var marker_active_1 = tslib_1.__importDefault(require("./marker-active"));
dependents_1.registerAction('selected-tooltip', selected_tooltip_1.SelectedTooltipAction);
dependents_1.registerInteraction('selected-tooltip', {
    start: [
        { trigger: dependents_1.VIEW_LIFE_CIRCLE.AFTER_PAINT, action: 'selected-tooltip:show' },
        { trigger: dependents_1.VIEW_LIFE_CIRCLE.AFTER_RENDER, action: 'selected-tooltip:show' },
        {
            trigger: 'plot:mousemove',
            action: 'selected-tooltip:show',
            throttle: { wait: 50, leading: true, trailing: false },
        },
        {
            trigger: 'plot:touchmove',
            action: 'selected-tooltip:show',
            throttle: { wait: 50, leading: true, trailing: false },
        },
    ],
});
dependents_1.registerAction('marker-active', marker_active_1.default);
dependents_1.registerInteraction('marker-active', {
    start: [
        { trigger: dependents_1.VIEW_LIFE_CIRCLE.AFTER_PAINT, action: 'marker-active:show' },
        { trigger: dependents_1.VIEW_LIFE_CIRCLE.AFTER_RENDER, action: 'marker-active:show' },
        {
            trigger: 'plot:mousemove',
            action: 'marker-active:show',
            throttle: { wait: 50, leading: true, trailing: false },
        },
        {
            trigger: 'plot:touchmove',
            action: 'marker-active:show',
            throttle: { wait: 50, leading: true, trailing: false },
        },
        {
            trigger: 'plot:mouseleave',
            action: 'marker-active:hide',
            throttle: { wait: 50, leading: true, trailing: false },
        },
        {
            trigger: 'plot:touchend',
            action: 'marker-active:hide',
            throttle: { wait: 50, leading: true, trailing: false },
        },
    ],
});
