// @antv/g2plot/lib/plots/percent-stacked-area/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../stacked-area/layer"));
var data_1 = require("../../util/data");
var PercentStackedAreaLayer = /** @class */ (function (_super) {
    tslib_1.__extends(PercentStackedAreaLayer, _super);
    function PercentStackedAreaLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'percentStackedArea';
        _this.baseType = 'stackedArea';
        return _this;
    }
    PercentStackedAreaLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            yAxis: {
                visible: true,
                label: {
                    visible: true,
                    formatter: function (v) {
                        var reg = /%/gi;
                        return v.replace(reg, '');
                    },
                },
            },
        });
    };
    PercentStackedAreaLayer.prototype.processData = function (originData) {
        var _a = this.options, xField = _a.xField, yField = _a.yField;
        return data_1.transformDataPercentage(originData, xField, [yField]);
    };
    PercentStackedAreaLayer.prototype.scale = function () {
        var metaConfig = {};
        var yField = this.options.yField;
        metaConfig[this.options.yField] = {
            tickCount: 6,
            alias: yField + " (%)",
            min: 0,
            max: 1,
            formatter: function (v) {
                var formattedValue = (v * 100).toFixed(1);
                return formattedValue + "%";
            },
        };
        this.options.meta = util_1.deepMix({}, metaConfig, this.options.meta);
        _super.prototype.scale.call(this);
    };
    return PercentStackedAreaLayer;
}(layer_1.default));
exports.default = PercentStackedAreaLayer;
global_1.registerPlotType('percentStackedArea', PercentStackedAreaLayer);
