// @antv/g2plot/esm/plots/percent-stacked-column/layer.js
import { __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import StackedColumn from '../stacked-column/layer';
import { transformDataPercentage } from '../../util/data';
var PercentStackedColumnLayer = /** @class */ (function (_super) {
    __extends(PercentStackedColumnLayer, _super);
    function PercentStackedColumnLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'percentStackedColumn';
        return _this;
    }
    PercentStackedColumnLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            label: {
                visible: true,
                position: 'middle',
                offset: 0,
            },
            yAxis: {
                visible: true,
                tick: {
                    visible: false,
                },
                grid: {
                    visible: false,
                },
                title: {
                    visible: true,
                },
                label: {
                    visible: false,
                },
            },
        });
    };
    PercentStackedColumnLayer.prototype.processData = function (originData) {
        var _a = this.options, xField = _a.xField, yField = _a.yField;
        return transformDataPercentage(originData || [], xField, [yField]);
    };
    PercentStackedColumnLayer.prototype.scale = function () {
        var metaConfig = {};
        var yField = this.options.yField;
        metaConfig[yField] = {
            tickCount: 6,
            alias: yField + " (%)",
            min: 0,
            max: 1,
            formatter: function (v) {
                var formattedValue = (v * 100).toFixed(1);
                return formattedValue + "%";
            },
        };
        this.options.meta = deepMix({}, this.options.meta, metaConfig);
        _super.prototype.scale.call(this);
    };
    return PercentStackedColumnLayer;
}(StackedColumn));
export default PercentStackedColumnLayer;
registerPlotType('percentStackedColumn', PercentStackedColumnLayer);
