// @antv/g2plot/esm/plots/gauge/layer.js
import { __extends } from "tslib";
/**
 * @author linhuiw
 * @description 仪表盘 layer
 */
import { deepMix, uniqueId } from '@antv/util';
import { registerPlotType } from '../../base/global';
import ViewLayer from '../../base/view-layer';
import { extractScale } from '../../util/scale';
import { GaugeShape } from './geometry/shape/gauge-shape';
import * as EventParser from './event';
var GaugeLayer = /** @class */ (function (_super) {
    __extends(GaugeLayer, _super);
    function GaugeLayer(props) {
        var _this = _super.call(this, props) || this;
        _this.type = 'gauge';
        return _this;
    }
    GaugeLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            startAngle: -7 / 6,
            endAngle: 1 / 6,
            rangeBackgroundStyle: {
                fill: '#f0f0f0',
            },
            rangeSize: 24,
            statistic: {
                position: ['50%', '80%'],
            },
            axis: {
                visible: true,
                offset: -10,
                tickCount: 21,
                subTickCount: 4,
                tickLine: {
                    visible: true,
                    length: 5,
                    style: {
                        stroke: '#aaa',
                        lineWidth: 2,
                    },
                },
                label: {
                    visible: true,
                    offset: 8,
                    style: {
                        fill: '#aaa',
                        fontSize: 16,
                        textAlign: 'center',
                        textBaseline: 'middle',
                    },
                },
            },
            pivot: {
                visible: true,
                thickness: 6,
                pin: {
                    visible: true,
                    size: 2,
                    style: {
                        fill: '#2E364B',
                    },
                },
                base: {
                    visible: true,
                    style: {
                        fill: '#EEEEEE',
                    },
                },
                pointer: {
                    visible: true,
                    style: {
                        fill: '#CFCFCF',
                    },
                },
            },
        });
    };
    GaugeLayer.prototype.init = function () {
        var _a = this.options, value = _a.value, range = _a.range;
        var rangeSorted = (range || []).map(function (d) { return +d; }).sort(function (a, b) { return a - b; });
        var _b = this.options, _c = _b.min, min = _c === void 0 ? rangeSorted[0] : _c, _d = _b.max, max = _d === void 0 ? rangeSorted[rangeSorted.length - 1] : _d, _e = _b.format, format = _e === void 0 ? function (d) { return "" + d; } : _e;
        var valueText = format(value);
        //const styleMix = this.getStyleMix();
        //this.options.styleMix = styleMix;
        this.options.data = [{ value: value || 0 }];
        this.options.valueText = valueText;
        this.options.min = min;
        this.options.max = max;
        this.options.format = format;
        this.initG2Shape();
        _super.prototype.init.call(this);
    };
    /**
     * 绘制指针
     */
    GaugeLayer.prototype.initG2Shape = function () {
        this.gaugeShape = new GaugeShape(uniqueId());
        this.gaugeShape.setOption(this.type, deepMix({}, this.options, {
            radius: 0.6,
            angle: 240,
            textPosition: '100%',
        }));
        this.gaugeShape.render();
    };
    GaugeLayer.prototype.geometryParser = function () {
        return 'gauge';
    };
    GaugeLayer.prototype.scale = function () {
        var _a = this.options, min = _a.min, max = _a.max, format = _a.format;
        var scales = {
            value: {},
        };
        extractScale(scales.value, {
            min: min,
            max: max,
            minLimit: min,
            maxLimit: max,
            nice: true,
            formatter: format,
            // 自定义 tick step
            tickInterval: 20,
        });
        // @ts-ignore
        this.setConfig('scales', scales);
        _super.prototype.scale.call(this);
    };
    GaugeLayer.prototype.coord = function () {
        var coordConfig = {
            type: 'polar',
            cfg: {
                radius: 1,
                startAngle: this.options.startAngle * Math.PI,
                endAngle: this.options.endAngle * Math.PI,
            },
        };
        this.setConfig('coordinate', coordConfig);
    };
    GaugeLayer.prototype.axis = function () {
        var axis = this.options.axis;
        var axesConfig = {
            value: {
                line: null,
                grid: null,
                tickLine: null,
            },
        };
        var offsetValue;
        if (axis.offset < 0) {
            offsetValue = axis.offset - this.options.rangeSize - axis.tickLine.length - axis.label.offset;
        }
        else {
            offsetValue = axis.offset + axis.tickLine.length + axis.label.offset;
        }
        if (axis.label.visible) {
            axesConfig.value.label = {
                offset: offsetValue,
                textStyle: axis.label.style,
                autoRotate: true,
            };
        }
        else {
            axesConfig.value.label = {
                style: {
                    opacity: 0,
                },
            };
        }
        axesConfig['1'] = false;
        axesConfig['value'] = axis.visible ? axesConfig['value'] : false;
        this.setConfig('axes', axesConfig);
    };
    GaugeLayer.prototype.addGeometry = function () {
        var pointerColor = this.options.pivot.pointer.style.fill || this.theme.defaultColor;
        var pointer = {
            type: 'point',
            position: {
                fields: ['value', '1'],
            },
            shape: {
                values: ['gauge'],
            },
            color: {
                values: [pointerColor],
            },
        };
        this.setConfig('geometry', pointer);
    };
    GaugeLayer.prototype.annotation = function () {
        var statistic = this.options.statistic;
        var annotationConfigs = [];
        // @ts-ignore
        if (statistic && statistic.visible) {
            var statistics = this.renderStatistic();
            annotationConfigs.push(statistics);
        }
        this.setConfig('annotations', annotationConfigs);
    };
    GaugeLayer.prototype.renderStatistic = function () {
        var containerSize = Math.max(this.options.width, this.options.height) / 20;
        var statistic = this.options.statistic;
        var text = {
            type: 'text',
            content: statistic.text,
            top: true,
            position: statistic.position,
            style: {
                fill: statistic.color,
                fontSize: statistic.size ? statistic.size : containerSize * 1.2,
                textAlign: 'center',
                textBaseline: 'middle',
            },
        };
        return text;
    };
    GaugeLayer.prototype.parseEvents = function () {
        _super.prototype.parseEvents.call(this, EventParser);
    };
    return GaugeLayer;
}(ViewLayer));
export default GaugeLayer;
registerPlotType('gauge', GaugeLayer);
