// @antv/g2plot/esm/plots/pie/component/label/base-label.js
import { __assign } from "tslib";
import { ext } from '@antv/matrix-util';
import { deepMix, isString } from '@antv/util';
import { getEndPoint, getLabelRotate, getAngleByPoint, getOverlapArea, near } from './utils';
import { getEllipsisText } from '../../../../util/text';
/** label text和line距离 4px */
export var CROOK_DISTANCE = 4;
export function percent2Number(value) {
    var percentage = Number(value.endsWith('%') ? value.slice(0, -1) : value);
    return percentage / 100;
}
/**
 * 超出panel边界的标签默认隐藏
 */
function checkInPanel(label, panel) {
    var box = label.getBBox();
    //  横向溢出 暂不隐藏
    if (!(panel.y <= box.y && panel.y + panel.height >= box.y + box.height)) {
        label.get('parent').set('visible', false);
    }
}
var PieBaseLabel = /** @class */ (function () {
    function PieBaseLabel(plot, cfg) {
        this.destroyed = false;
        this.plot = plot;
        this.coordinateBBox = this.plot.view.coordinateBBox;
        var options = deepMix(this.getDefaultOptions(), cfg, {});
        this.adjustOption(options);
        this.options = options;
        this.init();
    }
    /** 处理标签遮挡问题 */
    PieBaseLabel.prototype.adjustOverlap = function (labels, panel) {
        // clearOverlap;
        for (var i = 1; i < labels.length; i++) {
            var label = labels[i];
            var overlapArea = 0;
            for (var j = i - 1; j >= 0; j--) {
                var prev = labels[j];
                // fix: start draw point.x is error when textAlign is right
                var prevBox = prev.getBBox();
                var currBox = label.getBBox();
                // if the previous one is invisible, skip
                if (prev.get('parent').get('visible')) {
                    overlapArea = getOverlapArea(prevBox, currBox);
                    if (!near(overlapArea, 0)) {
                        label.get('parent').set('visible', false);
                        break;
                    }
                }
            }
        }
        labels.forEach(function (label) { return checkInPanel(label, panel); });
    };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    PieBaseLabel.prototype.adjustItem = function (item) {
        return;
    };
    PieBaseLabel.prototype.init = function () {
        var _this = this;
        this.container = this.getGeometry().labelsContainer;
        this.plot.view.on('beforerender', function () {
            _this.clear();
            _this.plot.canvas.draw();
        });
    };
    PieBaseLabel.prototype.render = function () {
        // 先清空 再重新渲染（避免双次绘制）
        this.clear();
        this.initArcPoints();
        this.drawTexts();
        this.drawLines();
    };
    PieBaseLabel.prototype.clear = function () {
        if (this.container) {
            this.container.clear();
        }
    };
    PieBaseLabel.prototype.hide = function () {
        this.container.set('visible', false);
        this.plot.canvas.draw();
    };
    PieBaseLabel.prototype.show = function () {
        this.container.set('visible', true);
        this.plot.canvas.draw();
    };
    PieBaseLabel.prototype.destroy = function () {
        if (this.container) {
            this.container.remove();
        }
        this.destroyed = true;
    };
    PieBaseLabel.prototype.getFilteredSum = function () {
        var angleField = this.plot.options.angleField;
        var filteredData = this.plot.view.filteredData || [];
        return filteredData.reduce(function (pre, filteredDataItem) {
            return pre + filteredDataItem[angleField];
        }, 0);
    };
    /** 绘制文本 */
    PieBaseLabel.prototype.drawTexts = function () {
        var _this = this;
        var _a = this.options, style = _a.style, formatter = _a.formatter, autoRotate = _a.autoRotate, offsetX = _a.offsetX, offsetY = _a.offsetY, adjustPosition = _a.adjustPosition, allowOverlap = _a.allowOverlap;
        var shapeInfos = this.getItems();
        var shapes = [];
        shapeInfos.map(function (shapeInfo, idx) {
            var attrs = deepMix({}, shapeInfo, style);
            var filteredSum = _this.getFilteredSum();
            var angleField = _this.plot.options.angleField;
            var percent = shapeInfo.origin[angleField] / filteredSum;
            var content = formatter
                ? formatter(shapeInfo.name, {
                    _origin: shapeInfo.origin,
                    percent: percent,
                }, idx)
                : shapeInfo.name;
            var itemGroup = _this.container.addGroup({
                name: 'itemGroup',
                index: idx,
                capture: false,
            });
            var textShape = itemGroup.addShape('text', {
                attrs: deepMix({}, attrs, {
                    x: shapeInfo.x + offsetX,
                    y: shapeInfo.y + offsetY,
                    text: content,
                }),
            });
            textShape.set('id', "text-" + shapeInfo.name + "-" + idx);
            shapes.push(textShape);
        });
        shapes.forEach(function (shape) {
            var panelBox = _this.coordinateBBox;
            _this.adjustText(shape, panelBox);
        });
        if (adjustPosition) {
            this.layout(shapes, shapeInfos, this.coordinateBBox);
        }
        if (!allowOverlap) {
            this.adjustOverlap(shapes, this.coordinateBBox);
        }
        shapes.forEach(function (label, idx) {
            if (autoRotate) {
                _this.rotateLabel(label, getLabelRotate(shapeInfos[idx].angle));
            }
        });
    };
    PieBaseLabel.prototype.adjustText = function (label, panelBox) {
        var box = label.getBBox();
        var width = box.width;
        var deltaWidth = 0;
        if (box.maxX > panelBox.maxX) {
            width = panelBox.maxX - box.minX;
        }
        else if (box.minX < panelBox.minX) {
            width = box.maxX - panelBox.minX;
        }
        if (label.attr('textAlign') === 'left') {
            label.attr('x', Math.max(box.x - deltaWidth, 0));
        }
        else if (label.attr('textAlign') === 'right') {
            label.attr('x', Math.max(box.maxX - deltaWidth, 0));
        }
        if (width !== box.width) {
            var font_1 = {};
            ['fontSize', 'fontFamily', 'fontWeight'].forEach(function (k) {
                font_1[k] = label.attr(k);
            });
            var ellipsisTexts = label
                .attr('text')
                .split('\n')
                .map(function (t) { return getEllipsisText(t, width, font_1); });
            label.attr('text', ellipsisTexts.join('\n'));
        }
    };
    /** 绘制拉线 */
    PieBaseLabel.prototype.drawLines = function () {
        var _this = this;
        if (this.options.line.visible) {
            var itemGroups = this.container.get('children');
            var center_1 = this.getCoordinate().center;
            itemGroups.forEach(function (labelGroup, idx) {
                var label = labelGroup.get('children')[0];
                var anchor = _this.arcPoints[idx];
                var inLeft = anchor.x < center_1.x;
                // 拉线 和 label 之间的距离
                var distance = _this.options.offset > 4 ? 4 : 0;
                var path = _this.getLinePath(label, anchor, distance);
                var style = _this.options.line;
                labelGroup.addShape('path', {
                    attrs: __assign({ path: path, stroke: anchor.color }, style),
                });
                // 由于拉线的存在 label 需要进行偏移
                label.attr('x', label.attr('x') + (inLeft ? -distance : distance));
            });
        }
    };
    /** 获取label leader-line, 默认 not smooth */
    PieBaseLabel.prototype.getLinePath = function (label, anchor, distance) {
        var smooth = this.options.line ? this.options.line.smooth : false;
        var angle = anchor.angle;
        var _a = this.getCoordinate(), center = _a.center, radius = _a.radius;
        var breakAt = getEndPoint(center, angle, radius + distance);
        if (distance < 4) {
            breakAt = anchor;
        }
        var inLeft = anchor.x < center.x;
        var box = label.getBBox();
        var labelPosition = { x: inLeft ? box.maxX : box.minX, y: box.y + box.height / 2 };
        var smoothPath = [
            'C',
            // 1st control point (of the curve)
            labelPosition.x +
                // 4 gives the connector a little horizontal bend
                (inLeft ? 1 : -1) * (distance < 4 ? distance / 2 : 4),
            labelPosition.y,
            2 * breakAt.x - anchor.x,
            2 * breakAt.y - anchor.y,
            breakAt.x,
            breakAt.y,
        ];
        var straightPath = ['L', /** pointy break */ breakAt.x, breakAt.y];
        var linePath = smooth ? smoothPath : straightPath;
        var path = ['M', labelPosition.x, labelPosition.y].concat(linePath).concat('L', anchor.x, anchor.y);
        return path.join(',');
    };
    PieBaseLabel.prototype.getGeometry = function () {
        return this.plot.view.geometries[0];
    };
    PieBaseLabel.prototype.getCoordinate = function () {
        var coordinate = this.getGeometry().coordinate;
        var center = coordinate.getCenter();
        // @ts-ignore
        var radius = coordinate.getRadius();
        var startAngle = coordinate.startAngle;
        return { center: center, radius: radius, startAngle: startAngle };
    };
    PieBaseLabel.prototype.adjustOption = function (options) {
        var offset = options.offset;
        var radius = this.getCoordinate().radius;
        if (isString(offset)) {
            offset = radius * percent2Number(offset);
        }
        options.offset = offset;
    };
    PieBaseLabel.prototype.rotateLabel = function (label, angle) {
        var x = label.attr('x');
        var y = label.attr('y');
        var matrix = ext.transform(label.getMatrix(), [
            ['t', -x, -y],
            ['r', angle],
            ['t', x, y],
        ]);
        label.setMatrix(matrix);
    };
    PieBaseLabel.prototype.getItems = function () {
        var _this = this;
        var offset = this.options.offset;
        var _a = this.getCoordinate(), center = _a.center, radius = _a.radius;
        var items = this.arcPoints.map(function (anchor) {
            var point = getEndPoint(center, anchor.angle, radius + offset);
            var item = __assign(__assign({}, anchor), point);
            _this.adjustItem(item);
            return item;
        });
        return items;
    };
    // 初始化圆弧上锚点
    PieBaseLabel.prototype.initArcPoints = function () {
        var _this = this;
        var angleField = this.plot.options.angleField;
        var elements = this.getGeometry().elements;
        var coord = this.getCoordinate();
        var center = coord.center, radius = coord.radius;
        var startAngle = this.getCoordinate().startAngle;
        // 位置信息使用 positionScale
        var positionScales = this.getGeometry().attributes.position.scales;
        var scale = positionScales[1];
        var scaleField = scale.field;
        var anchors = elements.map(function (ele) {
            var origin = ele.shape.get('origin');
            var color = origin.color;
            var originData = origin.data[0] || origin.data;
            var endAngle = startAngle + Math.PI * 2 * scale.scale(originData[scaleField]);
            var angle = (startAngle + endAngle) / 2;
            var point = getEndPoint(center, angle, radius);
            startAngle = endAngle;
            // 标签名 使用 angleField
            var name = scale.getText(originData[angleField]);
            var textAlign = point.x > center.x ? 'left' : 'right';
            return {
                x: point.x,
                y: point.y,
                color: color,
                name: name,
                origin: originData,
                // 实际的角度
                angle: getAngleByPoint(_this.getGeometry().coordinate, point),
                textAlign: textAlign,
            };
        });
        this.arcPoints = anchors;
    };
    return PieBaseLabel;
}());
export default PieBaseLabel;
