// @antv/g2plot/esm/plots/grouped-rose/layer.js
import { __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import RoseLayer from '../rose/layer';
var GroupedRoseLayer = /** @class */ (function (_super) {
    __extends(GroupedRoseLayer, _super);
    function GroupedRoseLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'groupedRose';
        _this.baseType = 'rose';
        return _this;
    }
    GroupedRoseLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            xAxis: {
                visible: true,
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                grid: {
                    visible: true,
                    alignTick: false,
                    style: {
                        lineWidth: 0.5,
                    },
                },
                label: {
                    offset: 5,
                    autoRotate: true,
                },
            },
            yAxis: {
                visible: false,
            },
        });
    };
    GroupedRoseLayer.prototype.getColorScale = function () {
        var groupField = this.options.groupField;
        if (groupField) {
            return this.view.getScaleByField(groupField);
        }
    };
    GroupedRoseLayer.prototype.adjustRoseAdjust = function () {
        return [
            {
                type: 'dodge',
                marginRatio: 1,
            },
        ];
    };
    GroupedRoseLayer.prototype.geometryTooltip = function () {
        this.rose.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.rose.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.rose.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.rose.tooltip.fields = [this.options.radiusField, this.options.categoryField, this.options.groupField];
            }
        }
    };
    return GroupedRoseLayer;
}(RoseLayer));
export default GroupedRoseLayer;
registerPlotType('groupedRose', GroupedRoseLayer);
