// @antv/g2plot/esm/util/view.js
import { head, filter, map, some } from '@antv/util';
import { ORIGIN } from '../dependents';
export function getAllGeometryByType(view, type) {
    return filter(view.geometries, function (geometry) { return geometry.type === type; });
}
export function getGeometryByType(view, type) {
    return head(getAllGeometryByType(view, type));
}
export function getGeometryShapes(geometry) {
    return map(geometry.elements, function (element) { return element.shape; });
}
export function forEachGeometry(view, callback) {
    view.geometries.forEach(callback);
}
export function forEachElement(view, callback) {
    view.geometries.forEach(function (geometry) {
        geometry.elements.forEach(function (element) {
            callback(element);
        });
    });
}
/** 检测是否有和已存在的Shape数据`相等`的情况 */
export function checkOriginEqual(cur, dones, compare) {
    return some(dones, function (done) {
        return compare(done.get(ORIGIN), cur.get(ORIGIN));
    });
}
/** 将label调整到panel内 */
export function moveInPanel(shape, panel) {
    var box = shape.getBBox();
    var deltaX = 0;
    var deltaY = 0;
    if (box.minX < panel.minX) {
        deltaX = box.minX - panel.minX;
    }
    else if (box.maxX > panel.maxX) {
        deltaX = box.maxX - panel.maxX;
    }
    if (deltaX) {
        shape.attr('x', shape.attr('x') - deltaX);
    }
    if (box.minY < panel.minY) {
        deltaY = box.minY - panel.minY;
    }
    else if (box.maxY > panel.maxY) {
        deltaY = box.maxY - panel.maxY;
    }
    if (deltaY) {
        shape.attr('y', shape.attr('y') - deltaY);
    }
}
/**
 * 计算两个矩形之间的堆叠区域面积
 */
export function getOverlapArea(a, b, margin) {
    if (margin === void 0) { margin = 0; }
    var xOverlap = Math.max(0, Math.min(a.x + a.width + margin, b.x + b.width + margin) - Math.max(a.x - margin, b.x - margin));
    var yOverlap = Math.max(0, Math.min(a.y + a.height + margin, b.y + b.height + margin) - Math.max(a.y - margin, b.y - margin));
    return xOverlap * yOverlap;
}
/** 检测是否和已布局的堆叠 */
export function checkShapeOverlap(cur, dones) {
    var box = cur.getBBox();
    return some(dones, function (done) {
        var target = done.getBBox();
        return getOverlapArea(box, target, 2) > 0;
    });
}
