// @antv/g2plot/lib/plots/compatiblePlots/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// 兼容 0.x的图表类型
var ring_1 = require("./ring");
Object.defineProperty(exports, "Ring", { enumerable: true, get: function () { return ring_1.default; } });
var group_column_1 = require("./group-column");
Object.defineProperty(exports, "GroupColumn", { enumerable: true, get: function () { return group_column_1.default; } });
var group_bar_1 = require("./group-bar");
Object.defineProperty(exports, "GroupBar", { enumerable: true, get: function () { return group_bar_1.default; } });
var percentage_stack_area_1 = require("./percentage-stack-area");
Object.defineProperty(exports, "PercentageStackArea", { enumerable: true, get: function () { return percentage_stack_area_1.default; } });
var percentage_stack_bar_1 = require("./percentage-stack-bar");
Object.defineProperty(exports, "PercentageStackBar", { enumerable: true, get: function () { return percentage_stack_bar_1.default; } });
var percentage_stack_column_1 = require("./percentage-stack-column");
Object.defineProperty(exports, "PercentageStackColumn", { enumerable: true, get: function () { return percentage_stack_column_1.default; } });
var stack_area_1 = require("./stack-area");
Object.defineProperty(exports, "StackArea", { enumerable: true, get: function () { return stack_area_1.default; } });
var stack_bar_1 = require("./stack-bar");
Object.defineProperty(exports, "StackBar", { enumerable: true, get: function () { return stack_bar_1.default; } });
var stack_column_1 = require("./stack-column");
Object.defineProperty(exports, "StackColumn", { enumerable: true, get: function () { return stack_column_1.default; } });
