// @antv/g2plot/esm/plots/percent-stacked-bar/layer.js
import { __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import StackedBar from '../stacked-bar/layer';
import { transformDataPercentage } from '../../util/data';
var PercentStackedBarLayer = /** @class */ (function (_super) {
    __extends(PercentStackedBarLayer, _super);
    function PercentStackedBarLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'percentStackedBar';
        return _this;
    }
    PercentStackedBarLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            xAxis: {
                visible: true,
                tickLine: {
                    visible: false,
                },
                grid: {
                    visible: false,
                },
                title: {
                    visible: true,
                    formatter: function (v) { return v + " (%)"; },
                },
                label: {
                    visible: false,
                    formatter: function (v) {
                        var reg = /%/gi;
                        return v.replace(reg, '');
                    },
                },
            },
        });
    };
    PercentStackedBarLayer.prototype.processData = function (originData) {
        var _a = this.options, xField = _a.xField, yField = _a.yField;
        var processData = _super.prototype.processData.call(this, originData);
        return transformDataPercentage(processData, yField, [xField]);
    };
    PercentStackedBarLayer.prototype.scale = function () {
        var metaConfig = {};
        var xField = this.options.xField;
        metaConfig[xField] = {
            tickCount: 6,
            alias: xField + " (%)",
            min: 0,
            max: 1,
            formatter: function (v) {
                var formattedValue = (v * 100).toFixed(1);
                return formattedValue + "%";
            },
        };
        this.options.meta = deepMix({}, this.options.meta, metaConfig);
        _super.prototype.scale.call(this);
    };
    return PercentStackedBarLayer;
}(StackedBar));
export default PercentStackedBarLayer;
registerPlotType('percentStackedBar', PercentStackedBarLayer);
