// @antv/g2plot/lib/plots/percent-stacked-column/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../stacked-column/layer"));
var data_1 = require("../../util/data");
var PercentStackedColumnLayer = /** @class */ (function (_super) {
    tslib_1.__extends(PercentStackedColumnLayer, _super);
    function PercentStackedColumnLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'percentStackedColumn';
        return _this;
    }
    PercentStackedColumnLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            label: {
                visible: true,
                position: 'middle',
                offset: 0,
            },
            yAxis: {
                visible: true,
                tick: {
                    visible: false,
                },
                grid: {
                    visible: false,
                },
                title: {
                    visible: true,
                },
                label: {
                    visible: false,
                },
            },
        });
    };
    PercentStackedColumnLayer.prototype.processData = function (originData) {
        var _a = this.options, xField = _a.xField, yField = _a.yField;
        return data_1.transformDataPercentage(originData || [], xField, [yField]);
    };
    PercentStackedColumnLayer.prototype.scale = function () {
        var metaConfig = {};
        var yField = this.options.yField;
        metaConfig[yField] = {
            tickCount: 6,
            alias: yField + " (%)",
            min: 0,
            max: 1,
            formatter: function (v) {
                var formattedValue = (v * 100).toFixed(1);
                return formattedValue + "%";
            },
        };
        this.options.meta = util_1.deepMix({}, this.options.meta, metaConfig);
        _super.prototype.scale.call(this);
    };
    return PercentStackedColumnLayer;
}(layer_1.default));
exports.default = PercentStackedColumnLayer;
global_1.registerPlotType('percentStackedColumn', PercentStackedColumnLayer);
