// @antv/g2plot/lib/plots/grouped-rose/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../rose/layer"));
var GroupedRoseLayer = /** @class */ (function (_super) {
    tslib_1.__extends(GroupedRoseLayer, _super);
    function GroupedRoseLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'groupedRose';
        _this.baseType = 'rose';
        return _this;
    }
    GroupedRoseLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            xAxis: {
                visible: true,
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                grid: {
                    visible: true,
                    alignTick: false,
                    style: {
                        lineWidth: 0.5,
                    },
                },
                label: {
                    offset: 5,
                    autoRotate: true,
                },
            },
            yAxis: {
                visible: false,
            },
        });
    };
    GroupedRoseLayer.prototype.getColorScale = function () {
        var groupField = this.options.groupField;
        if (groupField) {
            return this.view.getScaleByField(groupField);
        }
    };
    GroupedRoseLayer.prototype.adjustRoseAdjust = function () {
        return [
            {
                type: 'dodge',
                marginRatio: 1,
            },
        ];
    };
    GroupedRoseLayer.prototype.geometryTooltip = function () {
        this.rose.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.rose.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.rose.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.rose.tooltip.fields = [this.options.radiusField, this.options.categoryField, this.options.groupField];
            }
        }
    };
    return GroupedRoseLayer;
}(layer_1.default));
exports.default = GroupedRoseLayer;
global_1.registerPlotType('groupedRose', GroupedRoseLayer);
