// @antv/g2plot/lib/components/timeline/button.js
"use strict";
/**
 * @file 播放轴组件按钮
 * @author blackganglion
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var base_1 = tslib_1.__importDefault(require("../base"));
var Button = /** @class */ (function (_super) {
    tslib_1.__extends(Button, _super);
    function Button(cfg) {
        return _super.call(this, util_1.deepMix({}, cfg)) || this;
    }
    Button.prototype.renderInner = function (group) {
        var _this = this;
        this.initElement(group);
        this.updateElement();
        this.renderMarker();
        group.off('click');
        group.on('click', function () {
            _this.emit('click');
        });
    };
    Button.prototype.destroy = function () {
        this.group.off();
        _super.prototype.destroy.call(this);
    };
    Button.prototype.initElement = function (group) {
        this.circle = group.addShape('circle', {
            attrs: {
                x: this.config.x,
                y: this.config.y,
                r: this.config.r,
                fill: '#607889',
            },
        });
        this.startMarker = group.addShape('path', {
            attrs: {
                path: this.getStartMarkerPath(),
                fill: '#ffffff',
            },
        });
        this.pauseGroupMarker = group.addGroup();
        var width = (1 / 4) * this.config.r;
        var height = 0.5 * this.config.r * Math.sqrt(3);
        this.pauseLeftMarker = this.pauseGroupMarker.addShape('rect', {
            attrs: {
                x: this.config.x - (1 / 4 + 1 / 8) * this.config.r,
                y: this.config.y - height / 2,
                width: width,
                height: height,
                fill: '#ffffff',
            },
        });
        this.pauseRightMarker = this.pauseGroupMarker.addShape('rect', {
            attrs: {
                x: this.config.x + (1 / 8) * this.config.r,
                y: this.config.y - height / 2,
                width: width,
                height: height,
                fill: '#ffffff',
            },
        });
    };
    Button.prototype.updateElement = function () {
        this.circle.attr('x', this.config.x);
        this.circle.attr('y', this.config.y);
        this.circle.attr('r', this.config.r);
        this.startMarker.attr('path', this.getStartMarkerPath());
        var width = (1 / 4) * this.config.r;
        var height = 0.5 * this.config.r * Math.sqrt(3);
        this.pauseLeftMarker.attr('x', this.config.x - (1 / 4 + 1 / 8) * this.config.r);
        this.pauseLeftMarker.attr('y', this.config.y - height / 2);
        this.pauseLeftMarker.attr('width', width);
        this.pauseLeftMarker.attr('height', height);
        this.pauseRightMarker.attr('x', this.config.x + (1 / 8) * this.config.r);
        this.pauseRightMarker.attr('y', this.config.y - height / 2);
        this.pauseRightMarker.attr('width', width);
        this.pauseRightMarker.attr('height', height);
    };
    Button.prototype.renderMarker = function () {
        if (this.config.isPlay) {
            this.startMarker.hide();
            this.pauseGroupMarker.show();
        }
        else {
            this.startMarker.show();
            this.pauseGroupMarker.hide();
        }
    };
    /** 获取播放键 marker path */
    Button.prototype.getStartMarkerPath = function () {
        var sideLength = 0.5 * this.config.r * Math.sqrt(3);
        return [
            ['M', this.config.x - sideLength / Math.sqrt(3) / 2, this.config.y - sideLength / 2],
            ['L', this.config.x + sideLength / Math.sqrt(3), this.config.y],
            ['L', this.config.x - sideLength / Math.sqrt(3) / 2, this.config.y + sideLength / 2],
        ];
    };
    return Button;
}(base_1.default));
exports.default = Button;
