// @antv/g2plot/lib/interaction/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var core_1 = tslib_1.__importDefault(require("./core"));
var BaseInteraction = /** @class */ (function (_super) {
    tslib_1.__extends(BaseInteraction, _super);
    function BaseInteraction(cfg, viewLayer, interactionRange, interaction) {
        var _this = _super.call(this, cfg) || this;
        _this.viewLayer = viewLayer;
        _this.interactionRange = interactionRange;
        _this.interactionConfig = interaction;
        _this.disposables = [];
        return _this;
    }
    BaseInteraction.registerInteraction = function (type, ctor) {
        BaseInteraction.GLOBAL_INTERACTION_MAP[type] = ctor;
    };
    BaseInteraction.registerPlotInteraction = function (plotType, type, ctor) {
        if (!BaseInteraction.PLOT_INTERACTION_MAP[plotType]) {
            BaseInteraction.PLOT_INTERACTION_MAP[plotType] = {};
        }
        BaseInteraction.PLOT_INTERACTION_MAP[plotType][type] = ctor;
    };
    BaseInteraction.getInteraction = function (type, plotType) {
        if (plotType && BaseInteraction.PLOT_INTERACTION_MAP[plotType] && BaseInteraction[plotType][type]) {
            return BaseInteraction.PLOT_INTERACTION_MAP[plotType][type];
        }
        return BaseInteraction.GLOBAL_INTERACTION_MAP[type];
    };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    BaseInteraction.getInteractionRange = function (layerRange, interaction) {
        return undefined;
    };
    BaseInteraction.prototype.destroy = function () {
        util_1.each(this.disposables, function (fn) {
            fn();
        });
        this.disposables = [];
        this.clear();
        _super.prototype.destroy.call(this);
    };
    BaseInteraction.prototype.getViewLayer = function () {
        return this.viewLayer;
    };
    BaseInteraction.prototype.getRange = function () {
        return this.interactionRange;
    };
    BaseInteraction.prototype.getInteractionConfig = function () {
        return this.interactionConfig;
    };
    BaseInteraction.prototype.addDisposable = function (fn) {
        this.disposables.push(fn);
    };
    BaseInteraction.prototype.render = function () {
        return;
    };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    BaseInteraction.prototype.start = function (ev) {
        return;
    };
    BaseInteraction.GLOBAL_INTERACTION_MAP = {};
    BaseInteraction.PLOT_INTERACTION_MAP = {};
    return BaseInteraction;
}(core_1.default));
exports.default = BaseInteraction;
