// @antv/g2plot/lib/interaction/timeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var dependents_1 = require("../dependents");
var base_1 = tslib_1.__importDefault(require("./base"));
var bbox_1 = tslib_1.__importDefault(require("../util/bbox"));
var timeline_1 = tslib_1.__importDefault(require("../components/timeline"));
var DEFAULT_HEIGHT = 40;
function getValidTimeLineConfig(interaction) {
    return tslib_1.__assign({ loop: false, auto: true, height: DEFAULT_HEIGHT, padding: [0, 20, 0, 0], speed: 2 }, interaction);
}
var TimeLineInteraction = /** @class */ (function (_super) {
    tslib_1.__extends(TimeLineInteraction, _super);
    function TimeLineInteraction() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onChangeFn = util_1.throttle(_this.onChange.bind(_this), 20, { leading: true });
        return _this;
    }
    /** TimeLineInteraction new 时的范围参数 interactionRange */
    TimeLineInteraction.getInteractionRange = function (layerRange, interaction) {
        var config = getValidTimeLineConfig(interaction);
        var paddingTop = config.padding[0];
        var paddingBottom = config.padding[2];
        return new bbox_1.default(layerRange.minX, layerRange.maxY - config.height - paddingTop - paddingBottom, layerRange.width, config.height + paddingTop + paddingBottom);
    };
    TimeLineInteraction.prototype.setAnimate = function (isAnimate) {
        var geometries = this.view.geometries;
        this.view.animate(isAnimate);
        geometries.forEach(function (geom) {
            geom.animate(isAnimate);
        });
    };
    TimeLineInteraction.prototype.start = function () {
        return;
    };
    TimeLineInteraction.prototype.renderTimeLine = function () {
        var _this = this;
        this.config = getValidTimeLineConfig(this.getInteractionConfig());
        var viewRange = this.view.viewBBox;
        var _a = this.config, loop = _a.loop, padding = _a.padding, speed = _a.speed;
        var paddingTop = padding[0], paddingRight = padding[1], paddingBottom = padding[2], paddingLeft = padding[3];
        var range = this.getRange();
        var ticks = this.getTicks();
        var width = viewRange.width - paddingLeft - paddingRight;
        var timeLineConfig = {
            x: viewRange.minX + paddingLeft,
            y: range.tl.y + paddingTop,
            width: width,
            height: range.height - paddingTop - paddingBottom,
            loop: loop,
            ticks: ticks,
            speed: speed,
            defaultCurrentTick: ticks[0],
        };
        if (this.timeline) {
            if (!util_1.isEqual(timeLineConfig, this.timeLineConfig)) {
                this.timeLineConfig = timeLineConfig;
                this.timeline.update(timeLineConfig);
            }
        }
        else {
            this.container = this.canvas.addGroup();
            this.timeline = new timeline_1.default(tslib_1.__assign({ container: this.container }, timeLineConfig));
            this.timeline.init();
            this.timeline.render();
            this.timeline.on('timelinestart', function () {
                _this.originAnimation = _this.view.getOptions().animate;
                _this.setAnimate(true);
            });
            this.timeline.on('timelineend', function () {
                _this.setAnimate(_this.originAnimation);
            });
            this.timeline.on('timelinechange', this.onChangeFn);
            this.timeline.on('timelineupdate', this.onChange.bind(this));
            this.view.data(this.getFilterData(ticks[0]));
            this.timeLineConfig = timeLineConfig;
        }
    };
    TimeLineInteraction.prototype.onChange = function (tick) {
        var filterData = this.getFilterData(tick);
        this.view.changeData(filterData);
    };
    TimeLineInteraction.prototype.getFilterData = function (tick) {
        var field = this.config.field;
        var data = this.getViewLayer().options.data;
        return data.filter(function (item) { return item[field] === tick; });
    };
    TimeLineInteraction.prototype.getTicks = function () {
        var field = this.config.field;
        var data = this.getViewLayer().options.data;
        return util_1.uniq(data.map(function (item) { return item[field]; }));
    };
    /** 渲染 timeline */
    TimeLineInteraction.prototype.render = function () {
        var _this = this;
        this.firstRender = true;
        this.view.on(dependents_1.VIEW_LIFE_CIRCLE.BEFORE_RENDER, function () {
            _this.renderTimeLine();
        });
        this.view.on(dependents_1.VIEW_LIFE_CIRCLE.BEFORE_PAINT, function () {
            _this.renderTimeLine();
        });
        this.view.on(dependents_1.VIEW_LIFE_CIRCLE.AFTER_PAINT, function () {
            if (_this.config.auto && _this.firstRender) {
                _this.timeline.isPlay = true;
                _this.timeline.changePlayStatus();
            }
            _this.firstRender = false;
        });
        this.view.on(dependents_1.VIEW_LIFE_CIRCLE.AFTER_RENDER, function () {
            if (_this.config.auto && _this.firstRender) {
                _this.timeline.isPlay = true;
                _this.timeline.changePlayStatus();
            }
            _this.firstRender = false;
        });
    };
    TimeLineInteraction.prototype.clear = function () {
        if (this.timeline) {
            this.timeline.destroy();
            this.timeline = null;
        }
        if (this.container) {
            this.container.remove(true);
            this.container = null;
        }
    };
    return TimeLineInteraction;
}(base_1.default));
exports.default = TimeLineInteraction;
