// @antv/g2plot/esm/plots/stacked-column/layer.js
import { __assign, __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import ConnectedArea from '../../components/connected-area';
import BaseColumnLayer from '../column/layer';
import './theme';
import './component/label';
import './component/label-auto';
import { getGeometryByType } from '../../util/view';
var StackedColumnLayer = /** @class */ (function (_super) {
    __extends(StackedColumnLayer, _super);
    function StackedColumnLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'stackedColumn';
        _this.baseType = 'column';
        return _this;
    }
    StackedColumnLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            legend: {
                visible: true,
                position: 'right-top',
            },
            label: {
                visible: false,
                position: 'middle',
                offset: 0,
                adjustColor: true,
            },
            connectedArea: {
                visible: false,
                triggerOn: 'mouseenter',
            },
        });
    };
    StackedColumnLayer.prototype.init = function () {
        if (this.options.connectedArea.visible) {
            this.options.tooltip.crosshairs = null;
        }
        _super.prototype.init.call(this);
    };
    StackedColumnLayer.prototype.afterRender = function () {
        var props = this.options;
        // 绘制区域连接组件
        if (props.connectedArea.visible) {
            this.connectedArea = new ConnectedArea(__assign({ view: this.view, field: props.stackField, animation: props.animation === false ? false : true }, props.connectedArea));
        }
        _super.prototype.afterRender.call(this);
    };
    StackedColumnLayer.prototype.getColorScale = function () {
        var stackField = this.options.stackField;
        return this.view.getScaleByField(stackField);
    };
    StackedColumnLayer.prototype.adjustColumn = function (column) {
        column.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedColumnLayer.prototype.renderLabel = function () {
        var scales = this.config.scales;
        var _a = this.options, label = _a.label, yField = _a.yField;
        var scale = scales[yField];
        if (label && label.visible) {
            var geometry = getGeometryByType(this.view, 'interval');
            this.doRenderLabel(geometry, __assign({ type: 'stacked-column', formatter: scale.formatter && (function (value) { return scale.formatter(value); }) }, this.options.label));
        }
    };
    StackedColumnLayer.prototype.geometryTooltip = function () {
        this.column.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.column.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.column.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.column.tooltip.fields = [this.options.xField, this.options.yField, this.options.stackField];
            }
        }
    };
    return StackedColumnLayer;
}(BaseColumnLayer));
export default StackedColumnLayer;
registerPlotType('stackedColumn', StackedColumnLayer);
