// @antv/g2plot/lib/plots/range-bar/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../bar/layer"));
var label_1 = tslib_1.__importDefault(require("./component/label"));
var animation_1 = require("./animation");
var RangeBarLayer = /** @class */ (function (_super) {
    tslib_1.__extends(RangeBarLayer, _super);
    function RangeBarLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'rangeBar';
        return _this;
    }
    RangeBarLayer.getDefaultOptions = function () {
        return util_1.deepMix(_super.getDefaultOptions.call(this), {
            label: {
                visible: true,
                position: 'outer',
            },
            xAxis: {
                visible: true,
                autoRotateTitle: false,
                grid: {
                    visible: true,
                },
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                label: {
                    visible: true,
                    autoRotate: true,
                    autoHide: true,
                },
                title: {
                    visible: true,
                    spacing: 12,
                },
            },
            yAxis: {
                visible: true,
                autoRotateTitle: true,
                grid: {
                    visible: false,
                },
                line: {
                    visible: true,
                },
                tickLine: {
                    visible: true,
                },
                label: {
                    visible: true,
                    autoHide: true,
                    autoRotate: false,
                },
                title: {
                    visible: false,
                    spacing: 12,
                },
            },
        }, {});
    };
    RangeBarLayer.prototype.afterRender = function () {
        this.renderLabel();
        // 为更新动画缓存shape
        var shapeCaches = [];
        var geoms = this.view.geometries;
        util_1.each(geoms, function (geom) {
            var elements = geom.elements;
            util_1.each(elements, function (ele) {
                shapeCaches.push(ele.shape);
            });
        });
        animation_1.setShapeCache(shapeCaches);
        _super.prototype.afterRender.call(this);
    };
    RangeBarLayer.prototype.renderLabel = function () {
        if (this.options.label && this.options.label.visible) {
            var label = new label_1.default(tslib_1.__assign({ view: this.view, plot: this }, this.options.label));
            label.render();
        }
    };
    RangeBarLayer.prototype.animation = function () {
        _super.prototype.animation.call(this);
    };
    return RangeBarLayer;
}(layer_1.default));
exports.default = RangeBarLayer;
global_1.registerPlotType('rangeBar', RangeBarLayer);
