// @antv/g2plot/lib/plots/liquid/component/liquid-statistic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var statistic_1 = tslib_1.__importDefault(require("../../../components/statistic"));
var statistic_template_1 = require("./statistic-template");
var RingStatistic = /** @class */ (function (_super) {
    tslib_1.__extends(RingStatistic, _super);
    function RingStatistic(cfg) {
        var _this = _super.call(this, cfg) || this;
        _this.view = cfg.view;
        _this.plot = cfg.plot;
        _this.statisticClass = cfg.statisticClass;
        _this.adjustOptions();
        return _this;
    }
    /** 中心文本显示 */
    RingStatistic.prototype.getTotalHtmlString = function () {
        var statisticContainer = document.createElement('div');
        var _a = this.plot.options, _b = _a.value, value = _b === void 0 ? 0 : _b, statistic = _a.statistic;
        if (statistic && !statistic.visible) {
            return statisticContainer;
        }
        var htmlContent = this.options.htmlContent(value);
        if (util_1.isType(htmlContent, 'HTMLDivElement')) {
            statisticContainer.appendChild(htmlContent);
        }
        else {
            statisticContainer.innerHTML = statistic_template_1.getTemplate(htmlContent, this.statisticClass);
        }
        return statisticContainer;
    };
    RingStatistic.prototype.adjustOptions = function () {
        this.html = this.getTotalHtmlString();
        var _a = this.view.coordinateBBox, minX = _a.minX, minY = _a.minY, width = _a.width, height = _a.height;
        this.x = minX + width / 2;
        this.y = minY + height / 2;
    };
    return RingStatistic;
}(statistic_1.default));
exports.default = RingStatistic;
