// @antv/g2plot/lib/plots/grouped-column/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../column/layer"));
require("./theme");
var GroupedColumnLayer = /** @class */ (function (_super) {
    tslib_1.__extends(GroupedColumnLayer, _super);
    function GroupedColumnLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.baseType = 'column';
        _this.type = 'groupedColumn';
        return _this;
    }
    GroupedColumnLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            yAxis: {
                title: {
                    visible: true,
                },
            },
        });
    };
    GroupedColumnLayer.prototype.getResponsiveTheme = function () {
        return this.themeController.getResponsiveTheme('column');
    };
    GroupedColumnLayer.prototype.getColorScale = function () {
        var groupField = this.options.groupField;
        if (groupField) {
            return this.view.getScaleByField(groupField);
        }
    };
    GroupedColumnLayer.prototype.addGeometry = function () {
        _super.prototype.addGeometry.call(this);
    };
    GroupedColumnLayer.prototype.adjustColumn = function (column) {
        column.adjust = [
            {
                type: 'dodge',
                marginRatio: 0.1,
            },
        ];
    };
    GroupedColumnLayer.prototype.geometryTooltip = function () {
        this.column.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.column.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.column.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.column.tooltip.fields = [this.options.xField, this.options.yField, this.options.groupField];
            }
        }
    };
    return GroupedColumnLayer;
}(layer_1.default));
exports.default = GroupedColumnLayer;
global_1.registerPlotType('groupedColumn', GroupedColumnLayer);
