// @antv/g-gesture/lib/gestures/tap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var clock_1 = require("../utils/clock");
var gesture_1 = require("./gesture");
var TIME = 300;
var DISTANCE = 2;
var Tap = /** @class */ (function (_super) {
    tslib_1.__extends(Tap, _super);
    function Tap() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.EVENT = 'tap';
        return _this;
    }
    Tap.prototype.onTouchCancel = function (ev) { };
    Tap.prototype.onTouchEnd = function (ev) {
        var x = ev.x, y = ev.y;
        if (x - this.touchStartX < DISTANCE &&
            y - this.touchStartY < DISTANCE &&
            clock_1.clock.now() - this.touchStartTime < TIME) {
            this.emit({ x: x, y: y, event: ev });
        }
    };
    Tap.prototype.onTouchMove = function (ev) { };
    Tap.prototype.onTouchStart = function (ev) {
        var x = ev.x, y = ev.y;
        this.touchStartTime = clock_1.clock.now();
        this.touchStartX = x;
        this.touchStartY = y;
    };
    return Tap;
}(gesture_1.Gesture));
exports.Tap = Tap;
