// @antv/g2plot/esm/geoms/line/main.js
import { __extends } from "tslib";
import { get, isFunction, has, isString, isArray } from '@antv/util';
import ElementParser from '../base';
var LineParser = /** @class */ (function (_super) {
    __extends(LineParser, _super);
    function LineParser() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LineParser.prototype.init = function () {
        var props = this.plot.options;
        this.config = {
            type: 'line',
            position: {
                fields: [props.xField, props.yField],
            },
            connectNulls: props.connectNulls,
        };
        if (props.lineSize) {
            this.parseSize();
        }
        if (props.smooth) {
            this.config.shape = { values: ['smooth'] };
        }
        if (props.step) {
            this.config.shape = { values: [props.step] };
        }
        if (props.seriesField || props.color) {
            this.parseColor();
        }
        if (props.lineStyle || get(props, ['line', 'style'])) {
            this.parseStyle();
        }
    };
    LineParser.prototype.parseSize = function () {
        var sizeProps = this.plot.options.lineSize;
        var config = {};
        if (isFunction(sizeProps)) {
            config.callback = sizeProps;
        }
        else {
            config.values = [sizeProps];
        }
        this.config.size = config;
    };
    LineParser.prototype.parseColor = function () {
        var props = this.plot.options;
        var config = {};
        if (props.seriesField) {
            config.fields = [props.seriesField];
        }
        if (has(props, 'color')) {
            var color = props.color;
            if (isString(color)) {
                config.values = [color];
            }
            else if (isFunction(color)) {
                config.callback = color;
            }
            else if (isArray(color)) {
                if (props.seriesField) {
                    config.values = color;
                }
                else {
                    if (color.length > 0) {
                        config.values = [color[0]];
                    }
                }
            }
        }
        this.config.color = config;
    };
    LineParser.prototype.parseStyle = function () {
        var props = this.plot.options;
        var styleProps = props.lineStyle || get(props, ['line', 'style']);
        var config = {
            fields: null,
            callback: null,
            cfg: null,
        };
        if (isFunction(styleProps) && props.seriesField) {
            config.fields = [props.seriesField];
            config.callback = styleProps;
        }
        else {
            config.cfg = styleProps;
        }
        this.config.style = config;
    };
    return LineParser;
}(ElementParser));
export default LineParser;
