// @antv/g2plot/lib/components/label/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLabelComponent = exports.registerLabelComponent = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var base_1 = tslib_1.__importDefault(require("../base"));
var dependents_1 = require("../../dependents");
var bbox_1 = tslib_1.__importDefault(require("../../util/bbox"));
var LabelComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LabelComponent, _super);
    function LabelComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.labelsCfgMap = {};
        _this.lastLabelsCfgMap = {};
        return _this;
    }
    LabelComponent.prototype.getGeometry = function () {
        return this.geometry;
    };
    LabelComponent.prototype.getLabels = function () {
        return this.labels;
    };
    LabelComponent.prototype.show = function () {
        this.container.show();
    };
    LabelComponent.prototype.hide = function () {
        this.container.hide();
    };
    LabelComponent.prototype.initConfig = function (config) {
        var _this = this;
        this.layer = config.layer;
        var view = this.layer.view;
        this.view = view;
        this.geometry = config.geometry;
        this.coord = view.getCoordinate();
        this.options = util_1.deepMix(this.getDefaultOptions(), config.label);
        var callback = function () {
            _this.clear();
            _this.labelsCfgMap = {};
            _this.lastLabelsCfgMap = {};
            _this.render();
        };
        view.on(dependents_1.VIEW_LIFE_CIRCLE.AFTER_PAINT, callback);
        this.addDisposable(function () {
            view.off(dependents_1.VIEW_LIFE_CIRCLE.AFTER_PAINT, callback);
        });
    };
    LabelComponent.prototype.renderInner = function (group) {
        var _this = this;
        this.labels = [];
        this.labelsCfgMap = {};
        // 绘制 Label 图形
        util_1.each(this.geometry.elements, function (element, elementIdx) {
            var labels = [].concat(_this.drawLabelItem(group, element, elementIdx));
            util_1.each(labels, function (label, idx) {
                _this.adjustLabel(label, element, idx);
                if (!label.destroyed) {
                    _this.labels.push(label);
                    _this.labelsCfgMap[label.get('id')] = util_1.clone(label.attrs);
                }
            });
        });
        // 执行布局
        this.layoutLabels(this.geometry, this.labels);
        // 执行动画：参照 G2 Label 动画
        var lastLabelsCfgMap = this.lastLabelsCfgMap;
        var labelsCfgMap = this.labelsCfgMap;
        var animateCfg = this.geometry.animateOption ? dependents_1.getDefaultAnimateCfg('label', this.coord) : false;
        util_1.each(labelsCfgMap, function (attrs, id) {
            var shape = group.findById(id);
            if (shape) {
                if (lastLabelsCfgMap[id]) {
                    var oldAttrs = lastLabelsCfgMap[id];
                    // 图形发生更新
                    var updateAnimateCfg = util_1.get(animateCfg, 'update');
                    if (updateAnimateCfg) {
                        shape.stopAnimate();
                        shape.attr(oldAttrs);
                        dependents_1.doAnimate(shape, updateAnimateCfg, {
                            toAttrs: tslib_1.__assign({}, attrs),
                            coordinate: _this.coord,
                        });
                    }
                }
                else {
                    // 新生成的 shape
                    var appearAnimateCfg = util_1.get(animateCfg, 'appear');
                    if (appearAnimateCfg) {
                        shape.stopAnimate();
                        dependents_1.doAnimate(shape, appearAnimateCfg, {
                            toAttrs: tslib_1.__assign({}, shape.attr()),
                            coordinate: _this.coord,
                        });
                    }
                }
            }
            delete lastLabelsCfgMap[id];
        });
        util_1.each(lastLabelsCfgMap, function (attrs, id) {
            // 移除
            var leaveAnimateCfg = util_1.get(animateCfg, 'leave');
            if (leaveAnimateCfg) {
                var tempShape = group.addShape('text', {
                    attrs: attrs,
                    id: id,
                    name: 'label',
                });
                tempShape.stopAnimate();
                dependents_1.doAnimate(tempShape, leaveAnimateCfg, {
                    toAttrs: null,
                    coordinate: _this.coord,
                });
            }
        });
        this.lastLabelsCfgMap = this.labelsCfgMap;
    };
    LabelComponent.prototype.drawLabelText = function (group, attrs, extraCfgs) {
        if (extraCfgs === void 0) { extraCfgs = {}; }
        return group.addShape('text', tslib_1.__assign({ attrs: attrs }, extraCfgs));
    };
    LabelComponent.prototype.drawLabelItem = function (group, element, elementIndex) {
        var _this = this;
        var model = element.getModel();
        var items = [].concat(this.getLabelItemAttrs(element, elementIndex));
        var offset = this.getDefaultOffset();
        var offsetPoint = this.getLabelOffset();
        return util_1.map(items, function (attrs, index) {
            var _a;
            var position = {
                x: attrs.x + offsetPoint.x,
                y: attrs.y + offsetPoint.y,
            };
            var dataItem = util_1.isArray(model.mappingData) ? model.mappingData[index] : model.mappingData;
            var id = _this.getLabelId(dataItem);
            return _this.drawLabelText(group, tslib_1.__assign(tslib_1.__assign({}, attrs), position), (_a = {
                    id: id,
                    name: 'label',
                    offset: offset,
                    element: element
                },
                _a[dependents_1.ORIGIN] = dataItem,
                _a));
        });
    };
    /** 获取当前 Label 的 offset */
    LabelComponent.prototype.getDefaultOffset = function () {
        return Number(this.options.offset);
    };
    /** 默认实现：获取当前 Label 的 offset 点：包括 offset、offsetX、offsetY */
    LabelComponent.prototype.getLabelOffset = function () {
        var _a = this.options, offsetX = _a.offsetX, offsetY = _a.offsetY;
        return {
            x: util_1.isNumber(offsetX) ? offsetX : 0,
            y: util_1.isNumber(offsetY) ? offsetY : 0,
        };
    };
    /** 通过指定方向和系数获取整体 offset 点 */
    LabelComponent.prototype.getLabelOffsetByDimAndFactor = function (dim, factor) {
        var _a = this.options, offsetX = _a.offsetX, offsetY = _a.offsetY;
        var offset = this.getDefaultOffset();
        var offsetPoint = {
            x: 0,
            y: 0,
        };
        offsetPoint[dim] = offset * factor;
        if (util_1.isNumber(offsetX)) {
            offsetPoint.x += offsetX;
        }
        if (util_1.isNumber(offsetY)) {
            offsetPoint.y += offsetY;
        }
        return offsetPoint;
    };
    /** 初始化默认全局配置 */
    LabelComponent.prototype.getDefaultOptions = function () {
        return {};
    };
    /** 整理对所有 Labels 的布局调整 */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    LabelComponent.prototype.layoutLabels = function (geometry, labels) {
        // empty
    };
    LabelComponent.prototype.getLabelId = function (data) {
        var origin = data._origin;
        var type = this.geometry.type;
        var xScale = this.geometry.getXScale();
        var yScale = this.geometry.getYScale();
        var labelId = this.geometry.getElementId(data);
        if (type === 'line' || type === 'area') {
            // 折线图以及区域图，一条线会对应一组数据，即多个 labels，为了区分这些 labels，需要在 line id 的前提下加上 x 字段值
            labelId += " " + origin[xScale.field];
        }
        else if (type === 'path') {
            // path 路径图，无序，有可能存在相同 x 不同 y 的情况，需要通过 x y 来确定唯一 id
            labelId += " " + origin[xScale.field] + "-" + origin[yScale.field];
        }
        return labelId;
    };
    LabelComponent.prototype.getCoordinateBBox = function () {
        var coord = this.coord;
        var start = coord.start, end = coord.end;
        return new bbox_1.default(Math.min(start.x, end.x), Math.min(start.y, end.y), coord.getWidth(), coord.getHeight());
    };
    return LabelComponent;
}(base_1.default));
exports.default = LabelComponent;
// Label 组件注册
var LABEL_CONFIG_MAP = {};
function registerLabelComponent(type, component) {
    LABEL_CONFIG_MAP[type] = component;
}
exports.registerLabelComponent = registerLabelComponent;
function getLabelComponent(type) {
    return LABEL_CONFIG_MAP[type];
}
exports.getLabelComponent = getLabelComponent;
