// @antv/g2plot/esm/plots/stacked-rose/layer.js
import { __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import RoseLayer from '../rose/layer';
var StackedRoseLayer = /** @class */ (function (_super) {
    __extends(StackedRoseLayer, _super);
    function StackedRoseLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'stackedRose';
        _this.baseType = 'rose';
        return _this;
    }
    StackedRoseLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            xAxis: {
                visible: true,
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                grid: {
                    visible: true,
                    alignTick: false,
                    style: {
                        lineWidth: 0.5,
                    },
                },
                label: {
                    offset: 5,
                    autoRotate: true,
                },
            },
            yAxis: {
                visible: false,
            },
        });
    };
    StackedRoseLayer.prototype.getColorScale = function () {
        var stackField = this.options.stackField;
        return this.view.getScaleByField(stackField);
    };
    StackedRoseLayer.prototype.adjustRoseAdjust = function () {
        return [
            {
                type: 'stack',
            },
        ];
    };
    StackedRoseLayer.prototype.geometryTooltip = function () {
        this.rose.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.rose.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.rose.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.rose.tooltip.fields = [this.options.radiusField, this.options.categoryField, this.options.stackField];
            }
        }
    };
    return StackedRoseLayer;
}(RoseLayer));
export default StackedRoseLayer;
registerPlotType('stackedRose', StackedRoseLayer);
