// @antv/component/lib/annotation/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var line_1 = require("./line");
Object.defineProperty(exports, "Line", { enumerable: true, get: function () { return line_1.default; } });
var text_1 = require("./text");
Object.defineProperty(exports, "Text", { enumerable: true, get: function () { return text_1.default; } });
var arc_1 = require("./arc");
Object.defineProperty(exports, "Arc", { enumerable: true, get: function () { return arc_1.default; } });
var region_1 = require("./region");
Object.defineProperty(exports, "Region", { enumerable: true, get: function () { return region_1.default; } });
var image_1 = require("./image");
Object.defineProperty(exports, "Image", { enumerable: true, get: function () { return image_1.default; } });
var data_marker_1 = require("./data-marker");
Object.defineProperty(exports, "DataMarker", { enumerable: true, get: function () { return data_marker_1.default; } });
var data_region_1 = require("./data-region");
Object.defineProperty(exports, "DataRegion", { enumerable: true, get: function () { return data_region_1.default; } });
var region_filter_1 = require("./region-filter");
Object.defineProperty(exports, "RegionFilter", { enumerable: true, get: function () { return region_filter_1.default; } });
