// @antv/g2plot/lib/plots/range-column/animation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setShapeCache = void 0;
var dependents_1 = require("../../dependents");
var util_1 = require("@antv/util");
// 记录之前的状态
var shapeCache;
function clipInFromCenterVertical(shape, animateCfg) {
    var bbox = shape.getBBox();
    var centerY = bbox.minY + bbox.height / 2;
    shape.setClip({
        type: 'rect',
        attrs: {
            x: bbox.minX,
            y: centerY,
            width: bbox.width,
            height: 0,
        },
    });
    var cliper = shape.get('clipShape');
    cliper.animate({
        height: bbox.height,
        y: bbox.minY,
    }, animateCfg.duration, animateCfg.easing, function () {
        shape.setClip(null);
    }, animateCfg.delay);
}
clipInFromCenterVertical.animationName = 'clipInFromCenterVertical';
function setShapeCache(shapes) {
    shapeCache = shapes;
}
exports.setShapeCache = setShapeCache;
function updateFromCenterVertical(shape, animateCfg) {
    var fromPath = getShapeFromCache(shape).attr('path');
    var toPath = util_1.clone(shape.attr('path'));
    shape.attr('path', fromPath);
    shape.animate({
        path: toPath,
    }, animateCfg.duration, animateCfg.easing, animateCfg.callback, 100);
}
function getShapeFromCache(shape) {
    var id = shape.id;
    var target;
    util_1.each(shapeCache, function (s) {
        if (s.id === id) {
            target = s;
        }
    });
    return target;
}
updateFromCenterVertical.animationName = 'updateFromCenterVertical';
dependents_1.registerAnimation('clipInFromCenterVertical', clipInFromCenterVertical);
dependents_1.registerAnimation('updateFromCenterVertical', updateFromCenterVertical);
