// @antv/g2plot/lib/dependents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._ORIGIN = exports.ORIGIN = exports.TooltipCssConst = exports.HtmlTooltipTheme = exports.HtmlTooltip = exports.InteractionUtils = exports.GrammarInteraction = exports.GestureController = void 0;
var tslib_1 = require("tslib");
// dependents是基础依赖，以便按需使用plot
var gesture_1 = tslib_1.__importDefault(require("@antv/g2/lib/chart/controller/gesture"));
exports.GestureController = gesture_1.default;
var g2_1 = require("@antv/g2");
g2_1.registerComponentController('gesture', gesture_1.default);
// G
var g_base_1 = require("@antv/g-base");
Object.defineProperty(exports, "GraphicEvent", { enumerable: true, get: function () { return g_base_1.Event; } });
var g_canvas_1 = require("@antv/g-canvas");
Object.defineProperty(exports, "Canvas", { enumerable: true, get: function () { return g_canvas_1.Canvas; } });
var g_svg_1 = require("@antv/g-svg");
Object.defineProperty(exports, "SVG", { enumerable: true, get: function () { return g_svg_1.Canvas; } });
// G-Gesture
var g_gesture_1 = require("@antv/g-gesture");
Object.defineProperty(exports, "GM", { enumerable: true, get: function () { return g_gesture_1.GM; } });
Object.defineProperty(exports, "Wheel", { enumerable: true, get: function () { return g_gesture_1.Wheel; } });
// G2
var g2_2 = require("@antv/g2");
Object.defineProperty(exports, "View", { enumerable: true, get: function () { return g2_2.View; } });
Object.defineProperty(exports, "registerAnimation", { enumerable: true, get: function () { return g2_2.registerAnimation; } });
Object.defineProperty(exports, "registerGeometry", { enumerable: true, get: function () { return g2_2.registerGeometry; } });
Object.defineProperty(exports, "Geometry", { enumerable: true, get: function () { return g2_2.Geometry; } });
Object.defineProperty(exports, "Interaction", { enumerable: true, get: function () { return g2_2.Interaction; } });
Object.defineProperty(exports, "InteractionAction", { enumerable: true, get: function () { return g2_2.InteractionAction; } });
Object.defineProperty(exports, "registerInteraction", { enumerable: true, get: function () { return g2_2.registerInteraction; } });
Object.defineProperty(exports, "registerAction", { enumerable: true, get: function () { return g2_2.registerAction; } });
Object.defineProperty(exports, "registerShape", { enumerable: true, get: function () { return g2_2.registerShape; } });
Object.defineProperty(exports, "getTheme", { enumerable: true, get: function () { return g2_2.getTheme; } });
Object.defineProperty(exports, "Util", { enumerable: true, get: function () { return g2_2.Util; } });
Object.defineProperty(exports, "getShapeFactory", { enumerable: true, get: function () { return g2_2.getShapeFactory; } });
Object.defineProperty(exports, "ComponentController", { enumerable: true, get: function () { return g2_2.ComponentController; } });
Object.defineProperty(exports, "registerComponentController", { enumerable: true, get: function () { return g2_2.registerComponentController; } });
var constant_1 = require("@antv/g2/lib/constant");
Object.defineProperty(exports, "VIEW_LIFE_CIRCLE", { enumerable: true, get: function () { return constant_1.VIEW_LIFE_CIRCLE; } });
Object.defineProperty(exports, "COMPONENT_TYPE", { enumerable: true, get: function () { return constant_1.COMPONENT_TYPE; } });
Object.defineProperty(exports, "FIELD_ORIGIN", { enumerable: true, get: function () { return constant_1.FIELD_ORIGIN; } });
var tooltip_1 = require("@antv/g2/lib/chart/controller/tooltip");
Object.defineProperty(exports, "TooltipController", { enumerable: true, get: function () { return tooltip_1.default; } });
var marker_1 = require("@antv/g2/lib/util/marker");
Object.defineProperty(exports, "MarkerSymbols", { enumerable: true, get: function () { return marker_1.MarkerSymbols; } });
var grammar_interaction_1 = tslib_1.__importDefault(require("@antv/g2/lib/interaction/grammar-interaction"));
exports.GrammarInteraction = grammar_interaction_1.default;
var InteractionUtils = tslib_1.__importStar(require("@antv/g2/lib/interaction/action/util"));
exports.InteractionUtils = InteractionUtils;
var animate_1 = require("@antv/g2/lib/animate");
Object.defineProperty(exports, "DEFAULT_ANIMATE_CFG", { enumerable: true, get: function () { return animate_1.DEFAULT_ANIMATE_CFG; } });
Object.defineProperty(exports, "getDefaultAnimateCfg", { enumerable: true, get: function () { return animate_1.getDefaultAnimateCfg; } });
Object.defineProperty(exports, "doAnimate", { enumerable: true, get: function () { return animate_1.doAnimate; } });
var element_1 = require("@antv/g2/lib/geometry/element");
Object.defineProperty(exports, "Element", { enumerable: true, get: function () { return element_1.default; } });
// Component
var html_1 = tslib_1.__importDefault(require("@antv/component/lib/tooltip/html"));
exports.HtmlTooltip = html_1.default;
var html_theme_1 = tslib_1.__importDefault(require("@antv/component/lib/tooltip/html-theme"));
exports.HtmlTooltipTheme = html_theme_1.default;
var TooltipCssConst = tslib_1.__importStar(require("@antv/component/lib/tooltip/css-const"));
exports.TooltipCssConst = TooltipCssConst;
var component_1 = require("@antv/component");
Object.defineProperty(exports, "GroupComponent", { enumerable: true, get: function () { return component_1.GroupComponent; } });
Object.defineProperty(exports, "Axis", { enumerable: true, get: function () { return component_1.Axis; } });
Object.defineProperty(exports, "Legend", { enumerable: true, get: function () { return component_1.Legend; } });
Object.defineProperty(exports, "Tooltip", { enumerable: true, get: function () { return component_1.Tooltip; } });
Object.defineProperty(exports, "Slider", { enumerable: true, get: function () { return component_1.Slider; } });
Object.defineProperty(exports, "Scrollbar", { enumerable: true, get: function () { return component_1.Scrollbar; } });
// Coordinate
var coord_1 = require("@antv/coord");
Object.defineProperty(exports, "Coordinate", { enumerable: true, get: function () { return coord_1.Coordinate; } });
// Common
exports.ORIGIN = 'origin';
exports._ORIGIN = '_origin';
