// @antv/g2plot/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.G2 = void 0;
var tslib_1 = require("tslib");
// 通用配置
tslib_1.__exportStar(require("./interface/config"), exports);
var layer_1 = require("./base/layer");
Object.defineProperty(exports, "Layer", { enumerable: true, get: function () { return layer_1.default; } });
var view_layer_1 = require("./base/view-layer");
Object.defineProperty(exports, "ViewLayer", { enumerable: true, get: function () { return view_layer_1.default; } });
var plot_1 = require("./base/plot");
Object.defineProperty(exports, "Base", { enumerable: true, get: function () { return plot_1.default; } });
// 图形
tslib_1.__exportStar(require("./plots"), exports);
var progress_1 = require("./sparkline/progress");
Object.defineProperty(exports, "Progress", { enumerable: true, get: function () { return progress_1.default; } });
var ring_progress_1 = require("./sparkline/ring-progress");
Object.defineProperty(exports, "RingProgress", { enumerable: true, get: function () { return ring_progress_1.default; } });
var tiny_column_1 = require("./sparkline/tiny-column");
Object.defineProperty(exports, "TinyColumn", { enumerable: true, get: function () { return tiny_column_1.default; } });
var tiny_area_1 = require("./sparkline/tiny-area");
Object.defineProperty(exports, "TinyArea", { enumerable: true, get: function () { return tiny_area_1.default; } });
var tiny_line_1 = require("./sparkline/tiny-line");
Object.defineProperty(exports, "TinyLine", { enumerable: true, get: function () { return tiny_line_1.default; } });
// 混合图形
var dual_line_1 = require("./combo/dual-line");
Object.defineProperty(exports, "DualLine", { enumerable: true, get: function () { return dual_line_1.default; } });
var column_line_1 = require("./combo/column-line");
Object.defineProperty(exports, "ColumnLine", { enumerable: true, get: function () { return column_line_1.default; } });
var groupedColumn_line_1 = require("./combo/groupedColumn-line");
Object.defineProperty(exports, "GroupedColumnLine", { enumerable: true, get: function () { return groupedColumn_line_1.default; } });
var stackedColumn_line_1 = require("./combo/stackedColumn-line");
Object.defineProperty(exports, "StackedColumnLine", { enumerable: true, get: function () { return stackedColumn_line_1.default; } });
// 主题
var theme_1 = require("./theme");
Object.defineProperty(exports, "registerTheme", { enumerable: true, get: function () { return theme_1.registerTheme; } });
Object.defineProperty(exports, "registerGlobalTheme", { enumerable: true, get: function () { return theme_1.registerGlobalTheme; } });
var constraints_1 = require("./util/responsive/constraints");
Object.defineProperty(exports, "registerResponsiveConstraint", { enumerable: true, get: function () { return constraints_1.registerResponsiveConstraint; } });
var rules_1 = require("./util/responsive/rules");
Object.defineProperty(exports, "registerResponsiveRule", { enumerable: true, get: function () { return rules_1.registerResponsiveRule; } });
var theme_2 = require("./util/responsive/theme");
Object.defineProperty(exports, "registerResponsiveTheme", { enumerable: true, get: function () { return theme_2.registerResponsiveTheme; } });
Object.defineProperty(exports, "getResponsiveTheme", { enumerable: true, get: function () { return theme_2.getResponsiveTheme; } });
var state_manager_1 = require("./util/state-manager");
Object.defineProperty(exports, "StateManager", { enumerable: true, get: function () { return state_manager_1.default; } });
// 导出 g2 上层需要使用
var G2 = tslib_1.__importStar(require("@antv/g2"));
exports.G2 = G2;
