// @antv/g2plot/esm/plots/stacked-area/layer.js
import { __assign, __extends } from "tslib";
import { deepMix, get, each } from '@antv/util';
import { registerPlotType } from '../../base/global';
import BaseArea from '../area/layer';
import { getPlotComponents } from './component';
import './component/label/area-point';
import './component/label/area-point-auto';
import './theme';
var StackedAreaLayer = /** @class */ (function (_super) {
    __extends(StackedAreaLayer, _super);
    function StackedAreaLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.plotComponents = [];
        _this.type = 'stackedArea';
        _this.baseType = 'area';
        return _this;
    }
    StackedAreaLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            label: {
                visible: false,
                type: 'area',
            },
        });
    };
    StackedAreaLayer.prototype.beforeInit = function () {
        var visible = get(this.options, ['label', 'visible']);
        var type = get(this.options, ['label', 'type']);
        var options = this.options;
        if (visible) {
            if (type === 'line') {
                options.lineLabel = this.options.label;
            }
            if (type === 'area') {
                options.areaLabel = this.options.label;
            }
        }
        _super.prototype.beforeInit.call(this);
    };
    StackedAreaLayer.prototype.getColorScale = function () {
        var stackField = this.options.stackField;
        return this.view.getScaleByField(stackField);
    };
    StackedAreaLayer.prototype.label = function () {
        var props = this.options;
        var label = props.label;
        if (label && label.visible === false) {
            if (this.line) {
                this.line.label = false;
            }
            if (this.point) {
                this.point.label = false;
            }
            this.area.label = false;
            return;
        }
    };
    StackedAreaLayer.prototype.adjustArea = function (ele) {
        ele.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedAreaLayer.prototype.adjustLine = function (ele) {
        ele.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedAreaLayer.prototype.adjustPoint = function (ele) {
        ele.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedAreaLayer.prototype.afterRender = function () {
        this.renderPlotComponents();
        this.options.responsive = false;
        _super.prototype.afterRender.call(this);
    };
    StackedAreaLayer.prototype.geometryTooltip = function () {
        this.area.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.area.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.area.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.area.tooltip.fields = [this.options.xField, this.options.yField, this.options.stackField];
            }
        }
    };
    StackedAreaLayer.prototype.renderPlotComponents = function () {
        var _this = this;
        var componentsType = ['areaLabel', 'lineLabel'];
        each(componentsType, function (t) {
            var cfg = __assign({ view: _this.view, plot: _this }, _this.options[t]);
            var component = getPlotComponents(_this, t, cfg);
            if (component) {
                component.render();
                _this.plotComponents.push(component);
            }
        });
    };
    return StackedAreaLayer;
}(BaseArea));
export default StackedAreaLayer;
registerPlotType('stackedArea', StackedAreaLayer);
