// @antv/g2plot/lib/plots/heatmap/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var plot_1 = tslib_1.__importDefault(require("../../base/plot"));
var layer_1 = tslib_1.__importDefault(require("./layer"));
var Heatmap = /** @class */ (function (_super) {
    tslib_1.__extends(Heatmap, _super);
    function Heatmap() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Heatmap.prototype.createLayers = function (props) {
        var layerProps = util_1.deepMix({}, props);
        layerProps.type = 'heatmap';
        _super.prototype.createLayers.call(this, layerProps);
    };
    Heatmap.prototype.changeShape = function (type) {
        var layer = this.layers[0];
        layer.changeShape(type);
    };
    Heatmap.prototype.mappingSize = function (field) {
        var layer = this.layers[0];
        layer.mappingSize(field);
    };
    Heatmap.prototype.disableMappingSize = function () {
        var layer = this.layers[0];
        layer.disableMappingSize();
    };
    Heatmap.prototype.getSizeScale = function () {
        var layer = this.layers[0];
        return layer.getSizeScale();
    };
    Heatmap.getDefaultOptions = layer_1.default.getDefaultOptions;
    return Heatmap;
}(plot_1.default));
exports.default = Heatmap;
