// @antv/g2plot/esm/plots/heatmap/index.js
import { __extends } from "tslib";
import { deepMix } from '@antv/util';
import BasePlot from '../../base/plot';
import HeatmapLayer from './layer';
var Heatmap = /** @class */ (function (_super) {
    __extends(Heatmap, _super);
    function Heatmap() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Heatmap.prototype.createLayers = function (props) {
        var layerProps = deepMix({}, props);
        layerProps.type = 'heatmap';
        _super.prototype.createLayers.call(this, layerProps);
    };
    Heatmap.prototype.changeShape = function (type) {
        var layer = this.layers[0];
        layer.changeShape(type);
    };
    Heatmap.prototype.mappingSize = function (field) {
        var layer = this.layers[0];
        layer.mappingSize(field);
    };
    Heatmap.prototype.disableMappingSize = function () {
        var layer = this.layers[0];
        layer.disableMappingSize();
    };
    Heatmap.prototype.getSizeScale = function () {
        var layer = this.layers[0];
        return layer.getSizeScale();
    };
    Heatmap.getDefaultOptions = HeatmapLayer.getDefaultOptions;
    return Heatmap;
}(BasePlot));
export default Heatmap;
