// @antv/g2plot/esm/plots/heatmap/layer.js
import { __assign, __extends } from "tslib";
import { deepMix, valuesOfKey, each, isObject, isFunction } from '@antv/util';
import { getScale } from '@antv/scale';
import { registerPlotType } from '../../base/global';
import ViewLayer from '../../base/view-layer';
import { getRectPath, getCirclePath, getCircleCurve } from './shape';
import { getPlotComponents } from './component';
var HeatmapLayer = /** @class */ (function (_super) {
    __extends(HeatmapLayer, _super);
    function HeatmapLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'heatmap';
        _this.gridSize = [];
        _this.plotComponents = [];
        return _this;
    }
    HeatmapLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            shapeType: 'rect',
            legend: {
                visible: true,
                position: 'right-center',
            },
            tooltip: {
                shared: false,
                showCrosshairs: false,
                showMarkers: false,
            },
            xAxis: {
                visible: true,
                gridAlign: 'center',
                grid: {
                    visible: true,
                },
                tickLine: {
                    visible: true,
                },
                line: {
                    visible: false,
                },
                label: {
                    visible: true,
                    autoHide: true,
                    autoRotate: true,
                },
            },
            yAxis: {
                visible: true,
                gridAlign: 'center',
                grid: {
                    visible: true,
                    align: 'center',
                },
                tickLine: {
                    visible: true,
                },
                label: {
                    autoHide: true,
                    autoRotate: false,
                },
            },
            color: ['#9ae3d5', '#66cdbb', '#e7a744', '#f1e066', '#f27664', '#e7c1a2'],
            label: {
                visible: true,
                adjustColor: true,
                adjustPosition: true,
                offset: 0,
                style: {
                    stroke: 'rgba(255,255,255,0)',
                    lineWidth: 0,
                },
            },
            interactions: [{ type: 'tooltip' }],
        });
    };
    HeatmapLayer.prototype.afterRender = function () {
        this.renderPlotComponents();
        _super.prototype.afterRender.call(this);
    };
    HeatmapLayer.prototype.changeShape = function (type) {
        if (this.options.shapeType === type) {
            return;
        }
        this.options.shapeType = type;
        if (type === 'rect') {
            var shapes = this.getShape();
            this.circleToRect(shapes);
        }
        else if (type === 'circle') {
            var shapes = this.getShape();
            this.rectToCircle(shapes);
        }
    };
    HeatmapLayer.prototype.mappingSize = function (field) {
        if (this.options.sizeField && this.options.sizeField === field) {
            return;
        }
        this.options.sizeField = field;
        // 创建scale
        var values = valuesOfKey(this.options.data, field);
        var min = Math.min.apply(Math, values);
        var max = Math.max.apply(Math, values);
        var LinearScale = getScale('linear');
        var scale = new LinearScale({
            min: min,
            max: max,
        });
        var shapes = this.getShape();
        if (this.options.shapeType === 'rect') {
            this.rectSizeMapping(shapes, scale, field);
        }
        else if (this.options.shapeType === 'circle') {
            this.circleSizeMapping(shapes, scale, field);
        }
    };
    HeatmapLayer.prototype.disableMappingSize = function () {
        var shapes = this.getShape();
        if (this.options.shapeType === 'rect') {
            this.rectDisableSizeMapping(shapes);
        }
        else if (this.options.shapeType === 'circle') {
            this.circleDisableSizeMapping(shapes);
        }
    };
    HeatmapLayer.prototype.destroy = function () {
        each(this.plotComponents, function (component) {
            component.destroy();
        });
        _super.prototype.destroy.call(this);
    };
    HeatmapLayer.prototype.getSizeScale = function () {
        var sizeField = this.options.sizeField;
        if (sizeField) {
            this.view.getScaleByField(sizeField);
        }
    };
    HeatmapLayer.prototype.geometryParser = function () {
        return '';
    };
    HeatmapLayer.prototype.coord = function () {
        return;
    };
    HeatmapLayer.prototype.legend = function () {
        this.setConfig('legends', false);
    };
    HeatmapLayer.prototype.addGeometry = function () {
        this.gridSize = this.getGridSize();
        var geomConfig;
        if (this.options.shapeType === 'rect') {
            geomConfig = this.addRect();
        }
        else {
            var circle = this.addCircle();
            geomConfig = circle;
        }
        if (this.options.shapeStyle) {
            var styleConfig = {};
            if (isFunction(this.options.shapeStyle)) {
                styleConfig.fields = [
                    this.options.colorField,
                    this.options.xField,
                    this.options.yField,
                    this.options.sizeField,
                ];
                styleConfig.callback = this.options.shapeStyle;
            }
            else if (isObject(this.options.shapeStyle)) {
                styleConfig.cfg = this.options.shapeStyle;
            }
            geomConfig.style = styleConfig;
        }
        if (this.options.tooltip && (this.options.tooltip.fields || this.options.tooltip.formatter)) {
            this.geometryTooltip(geomConfig);
        }
        this.setConfig('geometry', geomConfig);
    };
    HeatmapLayer.prototype.addRect = function () {
        // 如果用户设置了size，将size数值转换为[0,1]区间
        var size = [0.3, 0.9];
        if (this.options.shapeSize) {
            size[0] = this.options.shapeSize[0] / this.gridSize[0];
            size[1] = this.options.shapeSize[1] / this.gridSize[1];
        }
        var rect = {
            type: 'polygon',
            position: {
                fields: [this.options.xField, this.options.yField],
            },
            color: {
                fields: [this.options.colorField],
                values: this.options.color,
            },
            shape: {
                values: ['rect'],
            },
            label: false,
        };
        if (this.options.sizeField) {
            rect.size = {
                fields: [this.options.sizeField],
                values: size,
            };
        }
        else {
            rect.size = {
                values: [1],
            };
        }
        return rect;
    };
    HeatmapLayer.prototype.addCircle = function () {
        var size = [0.3, 0.9];
        if (this.options.shapeSize) {
            size = this.options.shapeSize;
        }
        else {
            size[0] = this.gridSize[0] * size[0] * 0.5;
            size[1] = this.gridSize[1] * size[1] * 0.5;
        }
        var circle = {
            type: 'point',
            position: {
                fields: [this.options.xField, this.options.yField],
            },
            color: {
                fields: [this.options.colorField],
                values: this.options.color,
            },
            shape: {
                values: ['curvePoint'],
            },
            label: false,
        };
        if (this.options.sizeField) {
            circle.size = {
                fields: [this.options.sizeField],
                values: size,
            };
        }
        else {
            circle.size = {
                values: [Math.min(this.gridSize[0], this.gridSize[1]) * 0.5 * 0.9],
            };
        }
        return circle;
    };
    HeatmapLayer.prototype.geometryTooltip = function (config) {
        config.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            config.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            config.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                config.tooltip.fields = [this.options.xField, this.options.yField];
                if (this.options.colorField) {
                    config.tooltip.fields.push(this.options.colorField);
                }
            }
        }
    };
    HeatmapLayer.prototype.getGridSize = function () {
        if (this.options.padding === 'auto') {
            return [0, 0];
        }
        else {
            var viewRange = this.getViewRange();
            var _a = this.options, padding = _a.padding, xField = _a.xField, yField = _a.yField, data = _a.data;
            var width = viewRange.width - padding[1] - padding[3];
            var height = viewRange.height - padding[0] - padding[2];
            var xCount = valuesOfKey(data, xField).length;
            var yCount = valuesOfKey(data, yField).length;
            return [width / xCount, height / yCount];
        }
    };
    HeatmapLayer.prototype.circleToRect = function (shapes) {
        var _this = this;
        var gridSize = this.gridSize;
        each(shapes, function (shape) {
            var _a = shape.get('origin'), x = _a.x, y = _a.y, size = _a.size;
            var sizeRatio = (size * 2) / Math.min(gridSize[0], gridSize[1]);
            if (!_this.options.sizeField) {
                sizeRatio = 1;
            }
            var curvePath = getCircleCurve(x, y, size);
            var rectPath = getRectPath(x, y, gridSize[0], gridSize[1], sizeRatio);
            shape.stopAnimate();
            shape.attr('path', curvePath);
            shape.animate({
                path: rectPath,
            }, 500, 'easeLinear');
        });
    };
    HeatmapLayer.prototype.rectToCircle = function (shapes) {
        var _this = this;
        each(shapes, function (shape) {
            var coord = shape.get('coord');
            var points = shape.get('origin').points;
            var ps = [];
            each(points, function (p) {
                ps.push(coord.convertPoint(p));
            });
            var bbox = shape.getBBox();
            var width = bbox.width;
            var height = bbox.height;
            var centerX = bbox.minX + width / 2;
            var centerY = bbox.minY + height / 2;
            var offsetRatio = _this.options.sizeField ? 1 : 0.9;
            var curvePath = getCircleCurve(centerX, centerY, (Math.min(width, height) / 2) * offsetRatio);
            var circlePath = getCirclePath(centerX, centerY, (Math.min(width, height) / 2) * offsetRatio);
            shape.stopAnimate();
            shape.animate({
                path: curvePath,
            }, 500, 'easeLinear', function () {
                shape.attr('path', circlePath);
            });
        });
    };
    HeatmapLayer.prototype.rectSizeMapping = function (shapes, scale, field) {
        each(shapes, function (shape) {
            var data = shape.get('origin').data;
            var ratio = 0.3 + scale.scale(data[field]) * 0.6;
            shape.get('origin').size = ratio;
            var bbox = shape.getBBox();
            var width = bbox.width;
            var height = bbox.height;
            var centerX = bbox.minX + width / 2;
            var centerY = bbox.minY + height / 2;
            var path = getRectPath(centerX, centerY, width, height, ratio);
            shape.stopAnimate();
            shape.animate({
                path: path,
            }, 500, 'easeLinear');
        });
    };
    HeatmapLayer.prototype.circleSizeMapping = function (shapes, scale, field) {
        each(shapes, function (shape) {
            var data = shape.get('origin').data;
            var ratio = 0.3 + scale.scale(data[field]) * 0.6;
            var _a = shape.get('origin'), x = _a.x, y = _a.y, size = _a.size;
            var path = getCirclePath(x, y, size * ratio);
            shape.get('origin').size = size * ratio;
            shape.stopAnimate();
            shape.animate({
                path: path,
            }, 500, 'easeLinear');
        });
    };
    HeatmapLayer.prototype.circleDisableSizeMapping = function (shapes) {
        var _this = this;
        this.options.sizeField = null;
        each(shapes, function (shape) {
            var _a = shape.get('origin'), x = _a.x, y = _a.y;
            var size = Math.min(_this.gridSize[0], _this.gridSize[1]) * 0.9;
            shape.get('origin').size = size / 2;
            var path = getCirclePath(x, y, size / 2);
            shape.stopAnimate();
            shape.animate({
                path: path,
            }, 500, 'easeLinear');
        });
    };
    HeatmapLayer.prototype.rectDisableSizeMapping = function (shapes) {
        var _this = this;
        this.options.sizeField = null;
        each(shapes, function (shape) {
            var bbox = shape.getBBox();
            var width = bbox.width;
            var height = bbox.height;
            var centerX = bbox.minX + width / 2;
            var centerY = bbox.minY + height / 2;
            var path = getRectPath(centerX, centerY, _this.gridSize[0], _this.gridSize[1], 1);
            shape.get('origin').size = 1;
            shape.stopAnimate();
            shape.animate({
                path: path,
            }, 500, 'easeLinear');
        });
    };
    HeatmapLayer.prototype.getShape = function () {
        var elements = this.view.geometries[0].elements;
        var shapes = [];
        each(elements, function (ele) {
            shapes.push(ele.shape);
        });
        return shapes;
    };
    HeatmapLayer.prototype.renderPlotComponents = function () {
        var _this = this;
        each(this.plotComponents, function (component) {
            component.destroy();
        });
        this.plotComponents = [];
        var componentsType = ['label', 'legend'];
        each(componentsType, function (t) {
            var cfg = __assign({ view: _this.view, plot: _this }, _this.options[t]);
            var component = getPlotComponents(_this, t, cfg);
            if (component) {
                component.render();
                _this.plotComponents.push(component);
            }
        });
    };
    return HeatmapLayer;
}(ViewLayer));
export default HeatmapLayer;
registerPlotType('heatmap', HeatmapLayer);
