// @antv/g2plot/lib/plots/bullet/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Y_FIELD = exports.X_FIELD = exports.STACK_FIELD = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var EventParser = tslib_1.__importStar(require("./event"));
var view_layer_1 = tslib_1.__importDefault(require("../../base/view-layer"));
var scale_1 = require("../../util/scale");
var factory_1 = require("../../components/factory");
var factory_2 = require("../../geoms/factory");
var global_1 = require("../../base/global");
var bulletRect_1 = tslib_1.__importDefault(require("./component/bulletRect"));
var bulletTarget_1 = tslib_1.__importDefault(require("./component/bulletTarget"));
require("./theme");
var G2_GEOM_MAP = {
    bullet: 'interval',
};
var PLOT_GEOM_MAP = {
    interval: 'bullet',
};
exports.STACK_FIELD = '$$stackField$$';
exports.X_FIELD = '$$xField$$';
exports.Y_FIELD = '$$yField$$';
var BulletLayer = /** @class */ (function (_super) {
    tslib_1.__extends(BulletLayer, _super);
    function BulletLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'bullet';
        return _this;
    }
    BulletLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            data: [],
            stackField: exports.STACK_FIELD,
            xField: exports.X_FIELD,
            yField: exports.Y_FIELD,
            rangeColors: ['rgba(91, 143, 249, 0.45)'],
            measureSize: 12,
            rangeSize: 2,
            markerSize: 2,
            markerColors: [],
            markerStyle: {
                width: 2,
                fill: '#5B8FF9',
                lineWidth: 0,
            },
            axis: {
                visible: false,
                position: 'before',
                tickCount: 6,
                formatter: function (text, idx) { return "" + idx; },
                style: {
                    fill: 'rgba(0, 0, 0, 0.25)',
                    textBaseline: 'middle',
                    textAlign: 'center',
                    fontSize: 12,
                    lineHeight: 16,
                },
                tickLine: {
                    visible: true,
                    lineWidth: 1,
                    stroke: '#FFF',
                    lineDash: [4, 2],
                },
            },
            xAxis: {
                visible: true,
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                label: {
                    visible: true,
                },
            },
            yAxis: {
                visible: false,
                nice: false,
            },
            tooltip: {
                visible: false,
                trigger: 'item',
                crosshairs: false,
            },
            label: {
                visible: true,
                offset: 4,
                style: {
                    fill: 'rgba(0, 0, 0, 0.45)',
                    stroke: '#fff',
                    lineWidth: 1,
                },
            },
        });
    };
    BulletLayer.prototype.afterRender = function () {
        _super.prototype.afterRender.call(this);
        this.view.removeInteraction('legend-filter');
    };
    BulletLayer.prototype.scale = function () {
        var options = this.options;
        var scales = {};
        /** 配置y-scale */
        scales[options.yField] = {};
        if (util_1.has(options, 'yAxis')) {
            scale_1.extractScale(scales[options.yField], options.yAxis);
        }
        /** 配置x-scale */
        scales[options.xField] = {
            type: 'cat',
        };
        if (util_1.has(options, 'xAxis')) {
            scale_1.extractScale(scales[options.xField], options.xAxis);
        }
        this.setConfig('scales', scales);
        _super.prototype.scale.call(this);
    };
    BulletLayer.prototype.getOptions = function (props) {
        var options = _super.prototype.getOptions.call(this, props);
        this.adjustOptions(options);
        return options;
    };
    BulletLayer.prototype.afterInit = function () {
        _super.prototype.afterInit.call(this);
        var options = this.options;
        var ranges = options.data.map(function (d) { return d.ranges; });
        var targets = options.data.map(function (d) { return d.targets; });
        this.bulletRect = new bulletRect_1.default(this.view, {
            ranges: ranges,
            rangeMax: options.rangeMax,
            yField: options.yField,
            rangeSize: options.rangeSize,
            rangeColors: options.rangeColors || [],
            axis: options.axis,
        });
        this.bulletTarget = new bulletTarget_1.default(this.view, {
            targets: targets,
            yField: options.yField,
            markerSize: options.markerSize,
            markerColors: options.markerColors || [],
            markerStyle: options.markerStyle,
        });
    };
    BulletLayer.prototype.geometryParser = function (dim, type) {
        if (dim === 'g2') {
            return G2_GEOM_MAP[type];
        }
        return PLOT_GEOM_MAP[type];
    };
    BulletLayer.prototype.coord = function () {
        this.setConfig('coordinate', {
            actions: [['transpose']],
        });
    };
    /** 自定义子弹图图例 */
    BulletLayer.prototype.legend = function () {
        var options = this.options;
        var markerColor = options.markerStyle.fill;
        var measureColors = options.measureColors || this.theme.colors;
        var items = [
            {
                name: '实际进度',
                value: '实际进度',
                marker: {
                    symbol: 'square',
                    style: {
                        fill: measureColors[0],
                    },
                },
            },
            {
                name: '目标值',
                value: '目标值',
                marker: {
                    symbol: 'line',
                    style: {
                        stroke: markerColor,
                        lineWidth: 2,
                    },
                },
            },
        ];
        var legendOptions = tslib_1.__assign({ custom: true, position: 'bottom', items: items }, options.legend);
        // @ts-ignore
        this.setConfig('legends', legendOptions);
    };
    BulletLayer.prototype.addGeometry = function () {
        var options = this.options;
        var bullet = factory_2.getGeom('interval', 'main', {
            positionFields: [options.xField, options.yField],
            plot: this,
        });
        bullet.adjust = [
            {
                type: 'stack',
            },
        ];
        if (options.label) {
            bullet.label = this.extractLabel();
        }
        this.bullet = bullet;
        this.setConfig('geometry', bullet);
    };
    BulletLayer.prototype.parseEvents = function () {
        _super.prototype.parseEvents.call(this, EventParser);
    };
    BulletLayer.prototype.extractLabel = function () {
        var options = this.options;
        var label = util_1.deepMix({}, options.label);
        if (label.visible === false) {
            return false;
        }
        var labelConfig = factory_1.getComponent('label', tslib_1.__assign({ plot: this, labelType: 'barLabel', fields: [options.yField] }, label));
        return labelConfig;
    };
    BulletLayer.prototype.adjustOptions = function (options) {
        options.barSize = options.measureSize || 12;
        this.adjustYAxisOptions(options);
    };
    BulletLayer.prototype.adjustYAxisOptions = function (options) {
        var values = [];
        options.data.forEach(function (d) { return values.push(d.measures.reduce(function (a, b) { return a + b; }, 0)); });
        values.push(options.rangeMax);
        options.yAxis.max = Math.max.apply([], values);
    };
    BulletLayer.prototype.processData = function (dataOptions) {
        var options = this.options;
        var data = [];
        dataOptions.forEach(function (dataItem, dataIdx) {
            var _a;
            for (var valueIdx = 0; valueIdx < dataItem.measures.length; valueIdx += 1) {
                var value = dataItem.measures[valueIdx];
                var xField = dataItem.title || "" + dataIdx;
                data.push((_a = {},
                    _a[options.xField] = xField,
                    _a[options.yField] = value,
                    _a[options.stackField] = "" + valueIdx,
                    _a));
            }
        });
        return data;
    };
    return BulletLayer;
}(view_layer_1.default));
exports.default = BulletLayer;
global_1.registerPlotType('bullet', BulletLayer);
