// @antv/g2plot/lib/plots/stacked-rose/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var layer_1 = tslib_1.__importDefault(require("../rose/layer"));
var StackedRoseLayer = /** @class */ (function (_super) {
    tslib_1.__extends(StackedRoseLayer, _super);
    function StackedRoseLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'stackedRose';
        _this.baseType = 'rose';
        return _this;
    }
    StackedRoseLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            xAxis: {
                visible: true,
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                grid: {
                    visible: true,
                    alignTick: false,
                    style: {
                        lineWidth: 0.5,
                    },
                },
                label: {
                    offset: 5,
                    autoRotate: true,
                },
            },
            yAxis: {
                visible: false,
            },
        });
    };
    StackedRoseLayer.prototype.getColorScale = function () {
        var stackField = this.options.stackField;
        return this.view.getScaleByField(stackField);
    };
    StackedRoseLayer.prototype.adjustRoseAdjust = function () {
        return [
            {
                type: 'stack',
            },
        ];
    };
    StackedRoseLayer.prototype.geometryTooltip = function () {
        this.rose.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.rose.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.rose.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.rose.tooltip.fields = [this.options.radiusField, this.options.categoryField, this.options.stackField];
            }
        }
    };
    return StackedRoseLayer;
}(layer_1.default));
exports.default = StackedRoseLayer;
global_1.registerPlotType('stackedRose', StackedRoseLayer);
