// @antv/g2plot/lib/plots/bubble/layer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var global_1 = require("../../base/global");
var EventParser = tslib_1.__importStar(require("../scatter/event"));
var layer_1 = tslib_1.__importDefault(require("../scatter/layer"));
require("./shape");
require("./theme");
var BubbleLayer = /** @class */ (function (_super) {
    tslib_1.__extends(BubbleLayer, _super);
    function BubbleLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'bubble';
        return _this;
    }
    BubbleLayer.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.getDefaultOptions.call(this), {
            // 直径 min 4px；max 64px
            pointSize: [2, 32],
            pointStyle: {
                stroke: null,
                strokeOpacity: 1,
                fillOpacity: 0.5,
            },
            label: {
                position: 'middle',
                style: {
                    stroke: '#fff',
                    lineWidth: 2,
                },
            },
            shape: 'bubble-point',
        });
    };
    BubbleLayer.prototype.getSizeScale = function () {
        var sizeField = this.options.sizeField;
        if (sizeField) {
            this.view.getScaleByField(sizeField);
        }
    };
    BubbleLayer.prototype.legend = function () {
        var _a;
        _super.prototype.legend.call(this);
        if (this.options.legend && this.options.legend.visible === false) {
            return;
        }
        this.setConfig('legends', (_a = {},
            _a[this.options.sizeField] = false,
            _a));
    };
    BubbleLayer.prototype.parseEvents = function () {
        _super.prototype.parseEvents.call(this, EventParser);
    };
    return BubbleLayer;
}(layer_1.default));
exports.default = BubbleLayer;
global_1.registerPlotType('bubble', BubbleLayer);
